{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cf828ed7-28b4-4536-97d8-e570f9dfcb91",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-28T03:22:57.182751Z",
     "start_time": "2024-02-28T03:22:57.173905Z"
    }
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import gzip\n",
    "import math\n",
    "import copy\n",
    "import numpy\n",
    "import pandas\n",
    "import matplotlib\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "\n",
    "from sklearn.linear_model import LinearRegression\n",
    "from scipy.ndimage import gaussian_filter\n",
    "\n",
    "\n",
    "\n",
    "samples = [\"HG00621\", \"HG00741\", \"HG01952\", \"HG01978\", \"HG03516\"]\n",
    "replicates = [1, 2]\n",
    "phases = [\"mat\", \"pat\"]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16bea4dd-1444-4131-aa45-d9d1bf59a7ff",
   "metadata": {
    "ExecuteTime": {
     "start_time": "2024-02-28T03:22:58.513725Z"
    },
    "is_executing": true
   },
   "outputs": [],
   "source": [
    "class FastaReader(object):\n",
    "\n",
    "    def __init__(self):\n",
    "        self.data = {}\n",
    "        return\n",
    "\n",
    "    def read(self, fastafile):\n",
    "        chrom = \"\"\n",
    "        seq = \"\"\n",
    "\n",
    "        for l in open(fastafile):\n",
    "            l = l.strip()\n",
    "\n",
    "            if l.startswith(\">\"):\n",
    "                if chrom != \"\":\n",
    "                    self.data[chrom] = seq\n",
    "\n",
    "                chrom = l[1:]\n",
    "                seq = \"\"\n",
    "\n",
    "            else:\n",
    "                seq += l.lower()\n",
    "\n",
    "        self.data[chrom] = seq\n",
    "\n",
    "        return\n",
    "\n",
    "    def chromosomes(self):\n",
    "        return set(self.data.keys())\n",
    "\n",
    "    def get(self, chrom, start, end, zero_based=False, include_end=True):\n",
    "        if not zero_based:\n",
    "            start -= 1\n",
    "            end -= 1\n",
    "        if include_end:\n",
    "            end += 1\n",
    "\n",
    "        return self.data[chrom][start:end]\n",
    "\n",
    "\n",
    "hg38_fasta_fp = f\"/scratch/wzhang/ref/linear/hg38/bismark/hg38.fasta\"\n",
    "hg38_fasta = FastaReader()\n",
    "hg38_fasta.read(hg38_fasta_fp)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3c65fb340b90eb34",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T05:02:56.255073Z",
     "start_time": "2024-02-27T04:57:35.899673Z"
    },
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "KeyboardInterrupt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "chroms = hg38_fasta.chromosomes()\n",
    "cpg_sites = {}\n",
    "\n",
    "for chrom in chroms:\n",
    "    chrom_short = chrom.split()[0]\n",
    "    seq = hg38_fasta.data[chrom].upper()\n",
    "    \n",
    "    cpg_sites_this_chrom = set()\n",
    "    for i in range(len(seq)-1):\n",
    "\n",
    "        if seq[i: i+2] in [\"CG\"]:\n",
    "            cpg_sites_this_chrom.add(i+1)\n",
    "    \n",
    "    cpg_sites[chrom_short] = cpg_sites_this_chrom\n",
    "    \n",
    "    \n",
    "    # print(chrom_short, len(seq), len(cpg_sites_this_chrom), len(cpg_sites_this_chrom)/len(seq))\n",
    "    \n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ff0727d6319e5487",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-20T15:52:10.763687Z",
     "start_time": "2024-02-20T15:51:46.402354Z"
    },
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": [
    "cpg_ref_fp = \"/scratch/wzhang/ref/linear/hg38/bismark/cpg_sites.bed\"\n",
    "cpg_ref_fh = open(cpg_ref_fp, \"w\")\n",
    "\n",
    "\n",
    "for chrom in sorted(cpg_sites.keys()):\n",
    "    for pos in sorted(cpg_sites[chrom]):\n",
    "        line = f\"{chrom}\\t{pos}\\t{pos+2}\\n\"\n",
    "        cpg_ref_fh.write(line)\n",
    "\n",
    "cpg_ref_fh.close()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "833c55178f5636c0",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fed4aa08-838f-44cf-89ef-7144f7928a8b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-26T00:12:25.177023Z",
     "start_time": "2024-02-25T23:33:20.043274Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HG00621 1\n",
      "HG00621 2\n",
      "HG00741 1\n",
      "HG00741 2\n",
      "HG01952 1\n",
      "HG01952 2\n",
      "HG01978 1\n",
      "HG01978 2\n",
      "HG03516 1\n",
      "HG03516 2\n"
     ]
    }
   ],
   "source": [
    "# change name\n",
    "\n",
    "\n",
    "# Generate PerCpG methylation\n",
    "# prepare methylC for both types\n",
    "Î©\n",
    "\n",
    "# bismark_hg38_fu \n",
    "# bismark_hg38_default\n",
    "run_para_folder = \"bismark_hg38_default\"\n",
    "\n",
    "for s in samples:\n",
    "    for r in replicates:\n",
    "        \n",
    "        print(s, r)\n",
    "        wd = f\"/scratch/wzhang/projects/ggWGBS/{run_para_folder}/{s}/{r}/\"\n",
    "\n",
    "        f1i = f\"{wd}/bismark.cov.gz\"\n",
    "        f1o = f\"{wd}/bismark.bg\"\n",
    "        \n",
    "        f2i = f\"{wd}/bismark.cov.gz.bismark.cov.gz\"\n",
    "        f2o1 = f\"{wd}/track.cytosine.methylc\"\n",
    "        f2o2 = f\"{wd}/track.cpg.methylc\"\n",
    "        \n",
    "        f2o1_fh = open(f2o1, \"w\")\n",
    "        f2o2_fh = open(f2o2, \"w\")\n",
    "        \n",
    "        \n",
    "        \n",
    "        data = {}\n",
    "        i = 1\n",
    "        for l in gzip.open(f2i, \"rt\"):\n",
    "            l = l.strip().split()\n",
    "            chrom, start, end, ml, met, unmet = l\n",
    "            \n",
    "            start = int(start)\n",
    "            met = int(met)\n",
    "            unmet = int(unmet)\n",
    "            ml = float(ml) /100\n",
    "            cov = met + unmet\n",
    "            \n",
    "            if chrom not in data:\n",
    "                data[chrom] = {}\n",
    "            \n",
    "            strand = \"+\"\n",
    "            if start not in cpg_sites[chrom]:\n",
    "                strand = \"-\"\n",
    "            \n",
    "            data[chrom][start] = (met, cov)\n",
    "\n",
    "            \n",
    "            f2o1_fh.write(f\"{chrom}\\t{start-1}\\t{start}\\tCG\\t{ml}\\t{strand}\\t{cov}\\n\")\n",
    "        \n",
    "        f2o1_fh.close()\n",
    "        \n",
    "        \n",
    "        \n",
    "        cpg_sites[chrom_short] = cpg_sites_this_chrom\n",
    "        \n",
    "        for chrom in cpg_sites:\n",
    "            for start in sorted(cpg_sites[chrom]):\n",
    "                if chrom not in data:\n",
    "                    continue\n",
    "                \n",
    "                c1 = (0, 0)\n",
    "                c2 = (0, 0)\n",
    "                \n",
    "                if start in data[chrom]:\n",
    "                    c1 = data[chrom][start]\n",
    "                    del data[chrom][start]\n",
    "                \n",
    "                if start+1 in data[chrom]:\n",
    "                    c2 = data[chrom][start+1]\n",
    "                    del data[chrom][start+1]\n",
    "                \n",
    "                met = c1[0] + c2[0]\n",
    "                cov = c1[1] + c2[1]\n",
    "                \n",
    "                if cov == 0:\n",
    "                    continue\n",
    "                \n",
    "                ml = met / cov\n",
    "                \n",
    "                f2o2_fh.write(f\"{chrom}\\t{start-1}\\t{start+1}\\tCG\\t{ml}\\t+\\t{cov}\\n\")\n",
    "        \n",
    "        \n",
    "\n",
    "        f2o2_fh.close()\n",
    "        \n",
    "\n",
    "            \n",
    "            \n",
    "            \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d8a3152fc4fce617",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-20T18:09:37.295055Z",
     "start_time": "2024-02-20T16:53:06.007137Z"
    },
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HG00621\n",
      "1\n",
      "2\n",
      "HG00741\n",
      "1\n",
      "2\n",
      "HG01952\n",
      "1\n",
      "2\n",
      "HG01978\n",
      "1\n",
      "2\n",
      "HG03516\n",
      "1\n",
      "2\n",
      "HG00621\n",
      "1\n",
      "2\n",
      "HG00741\n",
      "1\n",
      "2\n",
      "HG01952\n",
      "1\n",
      "2\n",
      "HG01978\n",
      "1\n",
      "2\n",
      "HG03516\n",
      "1\n",
      "2\n"
     ]
    }
   ],
   "source": [
    "# clean methylGrapher lifted result, and merge replicate\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "for s in samples:\n",
    "    print(s)\n",
    "    \n",
    "    \n",
    "    \n",
    "    for r in replicates:\n",
    "        print(r)\n",
    "\n",
    "        fin_mg2l = f\"/scratch/wzhang/projects/ggWGBS/data/methylgrapher_pan_lifted/{s}_{r}.methylc.gz\"\n",
    "        \n",
    "        data_mg2l = {}\n",
    "        for l in gzip.open(fin_mg2l, \"rt\"):\n",
    "            l = l.strip().split()\n",
    "            # print(l)\n",
    "            chrom, start, end, context, ml, strand, cov, xxx = l\n",
    "            \n",
    "            \n",
    "            start = int(start)\n",
    "            end = int(end)\n",
    "            ml = float(ml)\n",
    "            cov = int(cov)\n",
    "            \n",
    "            met = int(round(ml * cov))\n",
    "            \n",
    "            if chrom not in data_mg2l:\n",
    "                data_mg2l[chrom] = {}\n",
    "            \n",
    "            if start not in data_mg2l[chrom]:\n",
    "                data_mg2l[chrom][start] = [0, 0]\n",
    "            \n",
    "            data_mg2l[chrom][start][0] += met\n",
    "            data_mg2l[chrom][start][1] += cov\n",
    "        \n",
    "        \n",
    "    \n",
    "        fout_mg2l = f\"/scratch/wzhang/projects/ggWGBS/data/methylgrapher_pan_lifted/{s}_{r}.cytosine.methylc\"\n",
    "        fhout_mg2l = open(fout_mg2l, \"w\")\n",
    "        for chrom in sorted(data_mg2l.keys()):\n",
    "            for start in sorted(data_mg2l[chrom]):\n",
    "                met, cov = data_mg2l[chrom][start]\n",
    "                ml = met / cov\n",
    "                fhout_mg2l.write(f\"{chrom}\\t{start}\\t{start+1}\\tCG\\t{ml}\\t+\\t{cov}\\n\")\n",
    "        fhout_mg2l.close()\n",
    "    \n",
    "    \n",
    "\n",
    "\n",
    "for s in samples:\n",
    "    print(s)\n",
    "    \n",
    "    data_mg2l = {}\n",
    "    for r in replicates:\n",
    "        print(r)\n",
    "\n",
    "        fin_mg2l = f\"/scratch/wzhang/projects/ggWGBS/data/methylgrapher_pan_lifted/{s}_{r}.methylc.gz\"\n",
    "        for l in gzip.open(fin_mg2l, \"rt\"):\n",
    "            l = l.strip().split()\n",
    "            # print(l)\n",
    "            chrom, start, end, context, ml, strand, cov, xxx = l\n",
    "            \n",
    "            start = int(start)\n",
    "            end = int(end)\n",
    "            ml = float(ml)\n",
    "            cov = int(cov)\n",
    "            \n",
    "            met = int(round(ml * cov))\n",
    "            \n",
    "            if chrom not in data_mg2l:\n",
    "                data_mg2l[chrom] = {}\n",
    "            \n",
    "            if start not in data_mg2l[chrom]:\n",
    "                data_mg2l[chrom][start] = [0, 0]\n",
    "            \n",
    "            data_mg2l[chrom][start][0] += met\n",
    "            data_mg2l[chrom][start][1] += cov\n",
    "        \n",
    "        \n",
    "    \n",
    "    fout_mg2l = f\"/scratch/wzhang/projects/ggWGBS/data/methylgrapher_pan_lifted_merge_replicate/{s}.cytosine.methylc\"\n",
    "    fhout_mg2l = open(fout_mg2l, \"w\")\n",
    "    for chrom in sorted(data_mg2l.keys()):\n",
    "        for start in sorted(data_mg2l[chrom]):\n",
    "            met, cov = data_mg2l[chrom][start]\n",
    "            ml = met / cov\n",
    "            fhout_mg2l.write(f\"{chrom}\\t{start}\\t{start+1}\\tCG\\t{ml}\\t+\\t{cov}\\n\")\n",
    "    fhout_mg2l.close()\n",
    "    \n",
    "\n",
    "        \n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "559bc0ffcf6b05d8",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-26T00:58:00.681583Z",
     "start_time": "2024-02-26T00:12:25.172523Z"
    },
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HG00621\n",
      "1\n",
      "2\n",
      "HG00741\n",
      "1\n",
      "2\n",
      "HG01952\n",
      "1\n",
      "2\n",
      "HG01978\n",
      "1\n",
      "2\n",
      "HG03516\n",
      "1\n",
      "2\n"
     ]
    }
   ],
   "source": [
    "# Merge replicate for bismark (cytosine and cpg)\n",
    "\n",
    "for s in samples:\n",
    "    print(s)\n",
    "    \n",
    "    data_bs_cytosine = {}\n",
    "    data_bs_cpg = {}\n",
    "    \n",
    "    for r in replicates:\n",
    "        print(r)\n",
    "        \n",
    "        \n",
    "        fin_bs = f\"/scratch/wzhang/projects/ggWGBS/{run_para_folder}/{s}/{r}/track.cytosine.methylc\"\n",
    "        for l in open(fin_bs):\n",
    "            l = l.strip().split()\n",
    "            # print(l)\n",
    "            chrom, start, end, context, ml, strand, cov = l\n",
    "            \n",
    "            start = int(start)\n",
    "            end = int(end)\n",
    "            ml = float(ml)\n",
    "            cov = int(cov)\n",
    "            \n",
    "            met = int(round(ml * cov))\n",
    "            \n",
    "            if chrom not in data_bs_cytosine:\n",
    "                data_bs_cytosine[chrom] = {}\n",
    "            \n",
    "            if start not in data_bs_cytosine[chrom]:\n",
    "                data_bs_cytosine[chrom][start] = [0, 0]\n",
    "            \n",
    "            data_bs_cytosine[chrom][start][0] += met\n",
    "            data_bs_cytosine[chrom][start][1] += cov\n",
    "            \n",
    "        \n",
    "        fin_bs = f\"/scratch/wzhang/projects/ggWGBS/{run_para_folder}/{s}/{r}/track.cpg.methylc\"\n",
    "        for l in open(fin_bs):\n",
    "            l = l.strip().split()\n",
    "            # print(l)\n",
    "            chrom, start, end, context, ml, strand, cov = l\n",
    "            \n",
    "            start = int(start)\n",
    "            end = int(end)\n",
    "            ml = float(ml)\n",
    "            cov = int(cov)\n",
    "            \n",
    "            met = int(round(ml * cov))\n",
    "            \n",
    "            if chrom not in data_bs_cpg:\n",
    "                data_bs_cpg[chrom] = {}\n",
    "            \n",
    "            if start not in data_bs_cpg[chrom]:\n",
    "                data_bs_cpg[chrom][start] = [0, 0]\n",
    "            \n",
    "            data_bs_cpg[chrom][start][0] += met\n",
    "            data_bs_cpg[chrom][start][1] += cov\n",
    "    \n",
    "    fout_bs = f\"/scratch/wzhang/projects/ggWGBS/data/bismark_hg38_merge_replicate/{s}.cytosine.methylc\"\n",
    "    fhout_bs = open(fout_bs, \"w\")\n",
    "    for chrom in sorted(data_bs_cytosine.keys()):\n",
    "        for start in sorted(data_bs_cytosine[chrom]):\n",
    "            met, cov = data_bs_cytosine[chrom][start]\n",
    "            ml = met / cov\n",
    "            fhout_bs.write(f\"{chrom}\\t{start}\\t{start+1}\\tCG\\t{ml}\\t+\\t{cov}\\n\")\n",
    "    fhout_bs.close()\n",
    "    \n",
    "    \n",
    "    fout_bs = f\"/scratch/wzhang/projects/ggWGBS/data/bismark_hg38_merge_replicate/{s}.cpg.methylc\"\n",
    "    fhout_bs = open(fout_bs, \"w\")\n",
    "    for chrom in sorted(data_bs_cpg.keys()):\n",
    "        for start in sorted(data_bs_cpg[chrom]):\n",
    "            met, cov = data_bs_cpg[chrom][start]\n",
    "            ml = met / cov\n",
    "            fhout_bs.write(f\"{chrom}\\t{start}\\t{start+2}\\tCG\\t{ml}\\t+\\t{cov}\\n\")\n",
    "    fhout_bs.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4f7b9984f88e9373",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-19T21:45:04.411084Z",
     "start_time": "2024-02-19T21:17:06.697362Z"
    },
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HG00621\n",
      "1\n",
      "2\n",
      "HG00741\n",
      "1\n",
      "2\n",
      "HG01952\n",
      "1\n",
      "2\n",
      "HG01978\n",
      "1\n",
      "2\n",
      "HG03516\n",
      "1\n",
      "2\n"
     ]
    }
   ],
   "source": [
    "# Merge replicate for methylGrapher\n",
    "\n",
    "\n",
    "\n",
    "for s in samples:\n",
    "    print(s)\n",
    "    data_mg = {}\n",
    "    \n",
    "    for r in replicates:\n",
    "        print(r)\n",
    "        \n",
    "        fin_mg = f\"/scratch/wzhang/projects/ggWGBS/data/methylgrapher_pan/{s}_{r}.cytosine.tsv\"\n",
    "        for l in open(fin_mg):\n",
    "            l = l.strip().split()\n",
    "            segID, pos, strand, context, unmet, met, cov, ml = l\n",
    "            \n",
    "            if segID not in data_mg:\n",
    "                data_mg[segID] = {}\n",
    "            \n",
    "            if pos not in data_mg[segID]:\n",
    "                data_mg[segID][pos] = [0, 0]\n",
    "            \n",
    "            data_mg[segID][pos][0] += int(met)\n",
    "            data_mg[segID][pos][1] += int(cov)\n",
    "\n",
    "    \n",
    "    fout_mg = f\"/scratch/wzhang/projects/ggWGBS/data/methylgrapher_pan_merge_replicate/{s}.cytosine.methyl\"\n",
    "    fhout_mg = open(fout_mg, \"w\")\n",
    "    for segID in data_mg.keys():\n",
    "        for pos in sorted(data_mg[segID]):\n",
    "            met, cov = data_mg[segID][pos]\n",
    "            ml = met / cov\n",
    "            fhout_mg.write(f\"{segID}\\t{pos}\\t+\\tCG\\t{cov - met}\\t{met}\\t{cov}\\t{ml}\\n\")\n",
    "    fhout_mg.close()\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "400f4b08a38acf00",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": [
    "# ----------- Divider ------------\n",
    "\n",
    "# Prepare tracks\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b39d6f5e-02c4-4940-aab7-54fa14bfff89",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T04:11:41.788880Z",
     "start_time": "2024-02-27T04:11:41.783621Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00621/1//track.cytosine.methylc /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00621/1//track.cytosine.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00621/1//track.cytosine.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00621/1//track.cytosine.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00621/1//track.cytosine.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00621/1//track.cytosine.methylc\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00621/1//track.cpg.methylc /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00621/1//track.cpg.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00621/1//track.cpg.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00621/1//track.cpg.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00621/1//track.cpg.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00621/1//track.cpg.methylc\n",
      "\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00621/2//track.cytosine.methylc /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00621/2//track.cytosine.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00621/2//track.cytosine.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00621/2//track.cytosine.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00621/2//track.cytosine.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00621/2//track.cytosine.methylc\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00621/2//track.cpg.methylc /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00621/2//track.cpg.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00621/2//track.cpg.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00621/2//track.cpg.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00621/2//track.cpg.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00621/2//track.cpg.methylc\n",
      "\n",
      "\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00741/1//track.cytosine.methylc /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00741/1//track.cytosine.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00741/1//track.cytosine.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00741/1//track.cytosine.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00741/1//track.cytosine.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00741/1//track.cytosine.methylc\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00741/1//track.cpg.methylc /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00741/1//track.cpg.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00741/1//track.cpg.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00741/1//track.cpg.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00741/1//track.cpg.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00741/1//track.cpg.methylc\n",
      "\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00741/2//track.cytosine.methylc /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00741/2//track.cytosine.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00741/2//track.cytosine.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00741/2//track.cytosine.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00741/2//track.cytosine.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00741/2//track.cytosine.methylc\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00741/2//track.cpg.methylc /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00741/2//track.cpg.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00741/2//track.cpg.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00741/2//track.cpg.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00741/2//track.cpg.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG00741/2//track.cpg.methylc\n",
      "\n",
      "\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01952/1//track.cytosine.methylc /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01952/1//track.cytosine.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01952/1//track.cytosine.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01952/1//track.cytosine.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01952/1//track.cytosine.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01952/1//track.cytosine.methylc\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01952/1//track.cpg.methylc /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01952/1//track.cpg.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01952/1//track.cpg.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01952/1//track.cpg.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01952/1//track.cpg.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01952/1//track.cpg.methylc\n",
      "\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01952/2//track.cytosine.methylc /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01952/2//track.cytosine.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01952/2//track.cytosine.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01952/2//track.cytosine.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01952/2//track.cytosine.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01952/2//track.cytosine.methylc\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01952/2//track.cpg.methylc /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01952/2//track.cpg.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01952/2//track.cpg.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01952/2//track.cpg.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01952/2//track.cpg.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01952/2//track.cpg.methylc\n",
      "\n",
      "\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01978/1//track.cytosine.methylc /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01978/1//track.cytosine.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01978/1//track.cytosine.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01978/1//track.cytosine.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01978/1//track.cytosine.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01978/1//track.cytosine.methylc\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01978/1//track.cpg.methylc /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01978/1//track.cpg.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01978/1//track.cpg.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01978/1//track.cpg.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01978/1//track.cpg.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01978/1//track.cpg.methylc\n",
      "\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01978/2//track.cytosine.methylc /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01978/2//track.cytosine.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01978/2//track.cytosine.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01978/2//track.cytosine.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01978/2//track.cytosine.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01978/2//track.cytosine.methylc\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01978/2//track.cpg.methylc /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01978/2//track.cpg.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01978/2//track.cpg.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01978/2//track.cpg.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01978/2//track.cpg.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG01978/2//track.cpg.methylc\n",
      "\n",
      "\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG03516/1//track.cytosine.methylc /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG03516/1//track.cytosine.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG03516/1//track.cytosine.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG03516/1//track.cytosine.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG03516/1//track.cytosine.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG03516/1//track.cytosine.methylc\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG03516/1//track.cpg.methylc /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG03516/1//track.cpg.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG03516/1//track.cpg.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG03516/1//track.cpg.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG03516/1//track.cpg.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG03516/1//track.cpg.methylc\n",
      "\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG03516/2//track.cytosine.methylc /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG03516/2//track.cytosine.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG03516/2//track.cytosine.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG03516/2//track.cytosine.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG03516/2//track.cytosine.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG03516/2//track.cytosine.methylc\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG03516/2//track.cpg.methylc /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG03516/2//track.cpg.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG03516/2//track.cpg.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG03516/2//track.cpg.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG03516/2//track.cpg.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_hg38_default/HG03516/2//track.cpg.methylc\n"
     ]
    }
   ],
   "source": [
    "# Prepare tracks for hg38\n",
    "\n",
    "for s in samples:\n",
    "    for r in replicates:\n",
    "        \n",
    "        wd = f\"/scratch/wzhang/projects/ggWGBS/{run_para_folder}/{s}/{r}/\"\n",
    "        \n",
    "        f2o1 = f\"{wd}/track.cytosine.methylc\"\n",
    "        f2o2 = f\"{wd}/track.cpg.methylc\"\n",
    "        \n",
    "        for fo in [f2o1, f2o2]:\n",
    "            \n",
    "            cmd1 = f\"sort -k1,1 -k2,2n -o {fo} {fo}\"\n",
    "            cmd2 = f\"bgzip {fo}\"\n",
    "            cmd3 = f\"tabix -p bed {fo}.gz\"\n",
    "            cmd4 = f\"cat {fo}.gz | gunzip > {fo}\"\n",
    "            \n",
    "            for cmd in [cmd1, cmd2, cmd3, cmd4]:\n",
    "                print(cmd)\n",
    "                \n",
    "        print()\n",
    "    print()\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "534b617456890fb1",
   "metadata": {
    "ExecuteTime": {
     "start_time": "2024-02-27T04:10:41.860743Z"
    },
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": [
    "# Make symbolic links to the data folder\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "for s in samples:\n",
    "    for r in replicates:\n",
    "        \n",
    "        wd = f\"/scratch/wzhang/projects/ggWGBS/{run_para_folder}/{s}/{r}/\"\n",
    "        \n",
    "        f2o1 = f\"cytosine.methylc\"\n",
    "        f2o2 = f\"cpg.methylc\"\n",
    "        \n",
    "        for fo in [f2o1, f2o2]:\n",
    "            cmd = f\"ln -s {wd}/track.{fo} /scratch/wzhang/projects/ggWGBS/data/bismark_hg38/{s}_{r}.{fo}\"\n",
    "            \n",
    "            print(cmd)\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1cbe68ee1d84c2b9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T04:52:47.477200Z",
     "start_time": "2024-02-27T04:52:47.464252Z"
    },
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/1/mat//track.cytosine.methylc /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/1/mat//track.cytosine.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/1/mat//track.cytosine.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/1/mat//track.cytosine.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/1/mat//track.cytosine.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/1/mat//track.cytosine.methylc\n",
      "\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/1/mat//track.cpg.methylc /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/1/mat//track.cpg.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/1/mat//track.cpg.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/1/mat//track.cpg.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/1/mat//track.cpg.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/1/mat//track.cpg.methylc\n",
      "\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/1/pat//track.cytosine.methylc /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/1/pat//track.cytosine.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/1/pat//track.cytosine.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/1/pat//track.cytosine.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/1/pat//track.cytosine.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/1/pat//track.cytosine.methylc\n",
      "\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/1/pat//track.cpg.methylc /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/1/pat//track.cpg.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/1/pat//track.cpg.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/1/pat//track.cpg.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/1/pat//track.cpg.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/1/pat//track.cpg.methylc\n",
      "\n",
      "\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/2/mat//track.cytosine.methylc /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/2/mat//track.cytosine.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/2/mat//track.cytosine.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/2/mat//track.cytosine.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/2/mat//track.cytosine.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/2/mat//track.cytosine.methylc\n",
      "\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/2/mat//track.cpg.methylc /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/2/mat//track.cpg.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/2/mat//track.cpg.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/2/mat//track.cpg.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/2/mat//track.cpg.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/2/mat//track.cpg.methylc\n",
      "\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/2/pat//track.cytosine.methylc /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/2/pat//track.cytosine.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/2/pat//track.cytosine.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/2/pat//track.cytosine.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/2/pat//track.cytosine.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/2/pat//track.cytosine.methylc\n",
      "\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/2/pat//track.cpg.methylc /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/2/pat//track.cpg.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/2/pat//track.cpg.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/2/pat//track.cpg.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/2/pat//track.cpg.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_ind_default/HG00621/2/pat//track.cpg.methylc\n",
      "\n",
      "\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/1/mat//track.cytosine.methylc /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/1/mat//track.cytosine.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/1/mat//track.cytosine.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/1/mat//track.cytosine.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/1/mat//track.cytosine.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/1/mat//track.cytosine.methylc\n",
      "\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/1/mat//track.cpg.methylc /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/1/mat//track.cpg.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/1/mat//track.cpg.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/1/mat//track.cpg.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/1/mat//track.cpg.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/1/mat//track.cpg.methylc\n",
      "\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/1/pat//track.cytosine.methylc /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/1/pat//track.cytosine.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/1/pat//track.cytosine.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/1/pat//track.cytosine.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/1/pat//track.cytosine.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/1/pat//track.cytosine.methylc\n",
      "\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/1/pat//track.cpg.methylc /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/1/pat//track.cpg.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/1/pat//track.cpg.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/1/pat//track.cpg.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/1/pat//track.cpg.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/1/pat//track.cpg.methylc\n",
      "\n",
      "\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/2/mat//track.cytosine.methylc /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/2/mat//track.cytosine.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/2/mat//track.cytosine.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/2/mat//track.cytosine.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/2/mat//track.cytosine.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/2/mat//track.cytosine.methylc\n",
      "\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/2/mat//track.cpg.methylc /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/2/mat//track.cpg.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/2/mat//track.cpg.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/2/mat//track.cpg.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/2/mat//track.cpg.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/2/mat//track.cpg.methylc\n",
      "\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/2/pat//track.cytosine.methylc /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/2/pat//track.cytosine.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/2/pat//track.cytosine.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/2/pat//track.cytosine.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/2/pat//track.cytosine.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/2/pat//track.cytosine.methylc\n",
      "\n",
      "sort -k1,1 -k2,2n -o /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/2/pat//track.cpg.methylc /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/2/pat//track.cpg.methylc\n",
      "bgzip /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/2/pat//track.cpg.methylc\n",
      "tabix -p bed /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/2/pat//track.cpg.methylc.gz\n",
      "cat /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/2/pat//track.cpg.methylc.gz | gunzip > /scratch/wzhang/projects/ggWGBS/bismark_ind_fu/HG00621/2/pat//track.cpg.methylc\n"
     ]
    }
   ],
   "source": [
    "# Prepare tracks for individual samples\n",
    "\n",
    "for run_para_folder in [\"bismark_ind_default\", \"bismark_ind_fu\"]:\n",
    "\n",
    "    for s in samples:\n",
    "        for r in replicates:\n",
    "            for p in phases:\n",
    "            \n",
    "                wd = f\"/scratch/wzhang/projects/ggWGBS/{run_para_folder}/{s}/{r}/{p}/\"\n",
    "                # print(wd)\n",
    "                \n",
    "                \n",
    "                f2o1 = f\"{wd}/track.cytosine.methylc\"\n",
    "                f2o2 = f\"{wd}/track.cpg.methylc\"\n",
    "                \n",
    "                for fo in [f2o1, f2o2]:\n",
    "                    \n",
    "                    cmd1 = f\"sort -k1,1 -k2,2n -o {fo} {fo}\"\n",
    "                    cmd2 = f\"bgzip {fo}\"\n",
    "                    cmd3 = f\"tabix -p bed {fo}.gz\"\n",
    "                    cmd4 = f\"cat {fo}.gz | gunzip > {fo}\"\n",
    "                    for cmd in [cmd1, cmd2, cmd3, cmd4]:\n",
    "                        print(cmd)\n",
    "                    print()\n",
    "                    \n",
    "            print()\n",
    "        \n",
    "        break\n",
    "        print()\n",
    "      "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75c9955cb797e1cb",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": [
    "# ----------- Divider ------------\n",
    "\n",
    "# Debug code from this point"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42a7b35117886b8b",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c12f73a63060773",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82df464d47ec4818",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "770919a39b9019b4",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c804cfeac39c7b14",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T06:07:15.705374Z",
     "start_time": "2024-02-27T05:13:32.218597Z"
    },
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HG00621 1 mat bismark_ind_default\n",
      "HG00621 1 pat bismark_ind_default\n",
      "HG00621 2 mat bismark_ind_default\n",
      "HG00621 2 pat bismark_ind_default\n",
      "HG00621 1 mat bismark_ind_fu\n",
      "HG00621 1 pat bismark_ind_fu\n",
      "HG00621 2 mat bismark_ind_fu\n",
      "HG00621 2 pat bismark_ind_fu\n"
     ]
    }
   ],
   "source": [
    "class FastaReader(object):\n",
    "\n",
    "    def __init__(self):\n",
    "        self.data = {}\n",
    "        return\n",
    "\n",
    "    def read(self, fastafile):\n",
    "        chrom = \"\"\n",
    "        seq = \"\"\n",
    "\n",
    "        for l in open(fastafile):\n",
    "            l = l.strip()\n",
    "\n",
    "            if l.startswith(\">\"):\n",
    "                if chrom != \"\":\n",
    "                    self.data[chrom] = seq\n",
    "\n",
    "                chrom = l[1:]\n",
    "                seq = \"\"\n",
    "\n",
    "            else:\n",
    "                seq += l.lower()\n",
    "\n",
    "        self.data[chrom] = seq\n",
    "\n",
    "        return\n",
    "\n",
    "    def chromosomes(self):\n",
    "        return set(self.data.keys())\n",
    "\n",
    "    def get(self, chrom, start, end, zero_based=False, include_end=True):\n",
    "        if not zero_based:\n",
    "            start -= 1\n",
    "            end -= 1\n",
    "        if include_end:\n",
    "            end += 1\n",
    "\n",
    "        return self.data[chrom][start:end]\n",
    "\n",
    "\n",
    "\n",
    "def get_cpg_sites(fasta_fp):\n",
    "    \n",
    "    fasta_obj = FastaReader()\n",
    "    fasta_obj.read(fasta_fp)\n",
    "\n",
    "    chroms = fasta_obj.chromosomes()\n",
    "    cpg_sites = {}\n",
    "    \n",
    "    for chrom in chroms:\n",
    "        chrom_short = chrom.split()[0]\n",
    "        seq = fasta_obj.data[chrom].upper()\n",
    "        \n",
    "        cpg_sites_this_chrom = set()\n",
    "        for i in range(len(seq)-1):\n",
    "    \n",
    "            if seq[i: i+2] in [\"CG\"]:\n",
    "                cpg_sites_this_chrom.add(i+1)\n",
    "        \n",
    "        cpg_sites[chrom_short] = cpg_sites_this_chrom\n",
    "    return cpg_sites\n",
    " \n",
    " \n",
    " \n",
    " \n",
    " \n",
    " \n",
    "for run_para_folder in [\"bismark_ind_default\", \"bismark_ind_fu\"]:\n",
    "\n",
    "    for s in samples:\n",
    "        \n",
    "        mat_genome = f\"/scratch/wzhang/ref/linear/pan/{s}_mat/genome.fasta\"\n",
    "        pat_genome = f\"/scratch/wzhang/ref/linear/pan/{s}_pat/genome.fasta\"\n",
    "        \n",
    "        mat_cpg_sites = get_cpg_sites(mat_genome)\n",
    "        pat_cpg_sites = get_cpg_sites(pat_genome)\n",
    "        \n",
    "        \n",
    "        for r in replicates:\n",
    "            \n",
    "            for p in phases:\n",
    "            \n",
    "                print(s, r, p, run_para_folder)\n",
    "                \n",
    "                wd = f\"/scratch/wzhang/projects/ggWGBS/{run_para_folder}/{s}/{r}/{p}\"\n",
    "                cpg_sites = mat_cpg_sites\n",
    "                if p == \"pat\":\n",
    "                    cpg_sites = pat_cpg_sites\n",
    "\n",
    "        \n",
    "                f1i = f\"{wd}/bismark.cov.gz\"\n",
    "                f1o = f\"{wd}/bismark.bg\"\n",
    "                \n",
    "                f2i = f\"{wd}/bismark.cov.gz.bismark.cov.gz\"\n",
    "                f2o1 = f\"{wd}/track.cytosine.methylc\"\n",
    "                f2o2 = f\"{wd}/track.cpg.methylc\"\n",
    "                \n",
    "                f2o1_fh = open(f2o1, \"w\")\n",
    "                f2o2_fh = open(f2o2, \"w\")\n",
    "                \n",
    "                \n",
    "                data = {}\n",
    "                i = 1\n",
    "                for l in gzip.open(f2i, \"rt\"):\n",
    "                    l = l.strip().split()\n",
    "                    chrom, start, end, ml, met, unmet = l\n",
    "                    \n",
    "                    start = int(start)\n",
    "                    met = int(met)\n",
    "                    unmet = int(unmet)\n",
    "                    ml = float(ml) /100\n",
    "                    cov = met + unmet\n",
    "                    \n",
    "                    if chrom not in data:\n",
    "                        data[chrom] = {}\n",
    "                    \n",
    "                    strand = \"+\"\n",
    "                    if start not in cpg_sites[chrom]:\n",
    "                        strand = \"-\"\n",
    "                    \n",
    "                    data[chrom][start] = (met, cov)\n",
    "                    \n",
    "                    # TODO add this later\n",
    "                    f2o1_fh.write(f\"{chrom}\\t{start-1}\\t{start}\\tCG\\t{ml}\\t{strand}\\t{cov}\\n\")\n",
    "                \n",
    "                f2o1_fh.close()\n",
    "                \n",
    "                \n",
    "                \n",
    "                cpg_sites[chrom_short] = cpg_sites_this_chrom\n",
    "                \n",
    "                for chrom in cpg_sites:\n",
    "                    for start in sorted(cpg_sites[chrom]):\n",
    "                        if chrom not in data:\n",
    "                            continue\n",
    "                        \n",
    "                        c1 = (0, 0)\n",
    "                        c2 = (0, 0)\n",
    "                        \n",
    "                        if start in data[chrom]:\n",
    "                            c1 = data[chrom][start]\n",
    "                            del data[chrom][start]\n",
    "                        \n",
    "                        if start+1 in data[chrom]:\n",
    "                            c2 = data[chrom][start+1]\n",
    "                            del data[chrom][start+1]\n",
    "                        \n",
    "                        met = c1[0] + c2[0]\n",
    "                        cov = c1[1] + c2[1]\n",
    "                        \n",
    "                        if cov == 0:\n",
    "                            continue\n",
    "                        \n",
    "                        ml = met / cov\n",
    "                        \n",
    "                        f2o2_fh.write(f\"{chrom}\\t{start-1}\\t{start+1}\\tCG\\t{ml}\\t+\\t{cov}\\n\")\n",
    "                \n",
    "        \n",
    "                f2o2_fh.close()\n",
    "        \n",
    "        break\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2c6694af33eed1",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
