{
 "cells": [
  {
   "cell_type": "code",
   "id": "cf828ed7-28b4-4536-97d8-e570f9dfcb91",
   "metadata": {},
   "source": [
    "import os\n",
    "import re\n",
    "import sys\n",
    "import gzip\n",
    "import time\n",
    "import json\n",
    "import math\n",
    "import copy\n",
    "import numpy\n",
    "import pandas\n",
    "import seaborn\n",
    "import matplotlib\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "import scipy.stats\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from scipy.ndimage import gaussian_filter\n",
    "\n",
    "from sklearn.linear_model import LinearRegression\n",
    "from sklearn.decomposition import PCA\n",
    "\n",
    "\n",
    "# seaborn.set()\n",
    "matplotlib.style.use('default')\n",
    "plt.axis('off')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "samples = [\"HG00621\", \"HG00741\", \"HG01952\", \"HG01978\", \"HG03516\"]\n",
    "replicates = [1, 2]\n",
    "phases = [\"mat\", \"pat\"]\n",
    "\n",
    "\n",
    "min_cov = 5\n",
    "\n",
    "\n",
    "pipeline_name_short = [\"bs\", \"bc\", \"bm\", \"gb\", \"mg\"]\n",
    "pipeline_name_full = [\"Bismark\", \"BISCUIT\", \"bwa-meth\", \"gemBS\", \"methylGrapher\"]\n",
    "pipeline_full_name_lookup = dict(zip(pipeline_name_short, pipeline_name_full))\n",
    "pipeline_color_lookup = {\n",
    "    \"bs\": \"#0d1780\",\n",
    "    \"bc\": \"#76b5c5\",\n",
    "    \"bm\": \"#873e23\",\n",
    "    \"gb\": \"#b7a1f9\",\n",
    "    \"mg\": \"#f0aa4f\"\n",
    "}\n",
    "\n",
    "\n",
    "individual_experiment_tissue_label_short = {\n",
    "    'ENC001': {\n",
    "        1: 'spleen',\n",
    "        2: 'skin',\n",
    "        3: 'testis',\n",
    "        4: 'thyroid gland',\n",
    "        5: 'pancreas',\n",
    "        6: 'lung',\n",
    "    },\n",
    "    'ENC002': {},\n",
    "    'ENC003': {},\n",
    "    'ENC004': {},\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "16bea4dd-1444-4131-aa45-d9d1bf59a7ff",
   "metadata": {},
   "source": [
    "\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "7bddc24a97e989bf",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [
    "# Common tools\n",
    "\n",
    "\n",
    "def methylc_reader(fp):\n",
    "    res = {}\n",
    "    fh = None\n",
    "    \n",
    "    if fp.lower().endswith(\".gz\") or fp.lower().endswith(\".gzip\"):\n",
    "        fh = gzip.open(fp, \"rt\")\n",
    "    else:\n",
    "        fh = open(fp)\n",
    "    for l in fh:\n",
    "        l = l.strip().split(\"\\t\")\n",
    "        chrom, start, end, context, ml, strand, cov = l\n",
    "        start = int(start)\n",
    "        ml = float(ml)\n",
    "        cov = int(cov)\n",
    "        \n",
    "        met = int(round(ml * cov))\n",
    "        \n",
    "        if chrom not in res:\n",
    "            res[chrom] = {}\n",
    "        res[chrom][start] = (met, cov)\n",
    "    \n",
    "    fh.close()\n",
    "    return res\n",
    "\n",
    "\n",
    "def methyl_reader(fp):\n",
    "    res = {}\n",
    "    fh = open(fp)\n",
    "    for l in fh:\n",
    "        l = l.strip().split(\"\\t\")\n",
    "        segID, pos, strand, context, unmet, met, cov, ml = l\n",
    "        \n",
    "        pos = int(pos)\n",
    "        met = int(met)\n",
    "        cov = int(cov)\n",
    "        \n",
    "        if segID not in res:\n",
    "            res[segID] = {}\n",
    "        res[segID][pos] = (met, cov)\n",
    "        \n",
    "    fh.close()\n",
    "    \n",
    "    return res\n",
    "\n",
    "\n",
    "def methylc_path(aligner, level, sample, replicate):\n",
    "    assert aligner in pipeline_name_short\n",
    "    assert level in [\"cpg\", \"cytosine\"]\n",
    "    assert sample in samples\n",
    "    assert replicate in replicates\n",
    "\n",
    "    res = None\n",
    "    if aligner == \"bs\":\n",
    "        res = f\"/scratch/wzhang/projects/ggWGBS/bismark_hg38_default/{sample}/{replicate}/track.{level}.methylc.gz\"\n",
    "    elif aligner == \"bc\":\n",
    "        res = f\"/scratch/wzhang/projects/ggWGBS/biscuit_run/{sample}/{replicate}/track.{level}.methylc.gz\"\n",
    "    elif aligner == \"bm\":\n",
    "        res = f\"/scratch/wzhang/projects/ggWGBS/bwameth/{sample}/{replicate}/track.{level}.methylc.gz\"\n",
    "    elif aligner == \"gb\":\n",
    "        res = f\"/scratch/wzhang/projects/ggWGBS/gembs_run/run_merged/8tracks/{sample}_{replicate}/track.{level}.methylc.gz\"\n",
    "    elif aligner == \"mg\":\n",
    "        #assert level == \"cytosine\"\n",
    "        res = f\"/scratch/wzhang/projects/ggWGBS/methylgrapher_no_trim/surjection/{sample}_{replicate}.methylc.gz\"\n",
    "\n",
    "    # assert res is not None\n",
    "    # assert os.path.exists(res)\n",
    "    return res\n",
    "\n",
    "\n",
    "\n",
    "def methylc_getter(aligner, level, sample, replicate):\n",
    "    methylc_p = methylc_path(aligner, level, sample, replicate)\n",
    "    if not os.path.exists(methylc_p):\n",
    "        return None\n",
    "    return methylc_reader(methylc_p)\n",
    "\n",
    "for s in samples:\n",
    "    for r in replicates:\n",
    "        for p in pipeline_name_short:\n",
    "            for level in [\"cpg\", \"cytosine\"]:\n",
    "                methylc_p = methylc_path(p, level, s, r)\n",
    "                if not os.path.exists(methylc_p):\n",
    "                    print(\"ERROR\", s, r, p, level, methylc_p)\n",
    "\n",
    "\n",
    "\n",
    "        "
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "1761add8fb2418f8",
   "metadata": {
    "collapsed": false,
    "is_executing": true,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [
    "# Just reading in necessary data\n",
    "\n",
    "bismark_cpg_data = {}\n",
    "methylgrapher_lifted_cytosine_data = {}\n",
    "methylgrapher_cpg_data = {}\n",
    "\n",
    "\n",
    "\n",
    "for s in samples:\n",
    "    bismark_cpg_data[s] = {}\n",
    "    methylgrapher_lifted_cytosine_data[s] = {}\n",
    "    methylgrapher_cpg_data[s] = {}\n",
    "    \n",
    "    for r in replicates:\n",
    "        print(s, r)\n",
    "        \n",
    "        \n",
    "        \n",
    "        bs_fp = f\"/scratch/wzhang/projects/ggWGBS/data/bismark_hg38/{s}_{r}.cpg.methylc\"\n",
    "        mg_fp = f\"/scratch/wzhang/projects/ggWGBS/methylgrapher_no_trim/{s}/{r}/mq10.graph.CG.merged.tsv\"\n",
    "        \n",
    "        mg2l_fp = f\"/scratch/wzhang/projects/ggWGBS/methylgrapher_no_trim/surjection/{s}_{r}.methylc.gz\"\n",
    "        \n",
    "        d = {}\n",
    "        for l in open(mg_fp):\n",
    "            l = l.strip().split(\"\\t\")\n",
    "            cpg_ID, met, cov = l\n",
    "            met = int(met)\n",
    "            cov = int(cov)\n",
    "            d[cpg_ID] = (met, cov)\n",
    "            \n",
    "            \n",
    "        methylgrapher_cpg_data[s][r] = d\n",
    "        \n",
    "\n",
    "        \n",
    "        # bismark_cpg_data[s][r] = methylc_reader(bs_fp)\n",
    "        methylgrapher_lifted_cytosine_data[s][r] = methylc_reader(mg2l_fp)\n",
    "        \n",
    "        \n",
    "\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "d519648a-9dc5-4233-98e0-0dd093e1fb3c",
   "metadata": {},
   "source": [
    "# Just reading in necessary data for Figure 5A\n",
    "\n",
    "bismark_cpg_data = {}\n",
    "methylgrapher_cytosine_data = {}\n",
    "\n",
    "\n",
    "\n",
    "for s in samples:\n",
    "    bismark_cpg_data[s] = {}\n",
    "    methylgrapher_cytosine_data[s] = {}\n",
    "    \n",
    "    for r in replicates:\n",
    "        print(s, r)\n",
    "\n",
    "        bs_fp = f\"/scratch/wzhang/projects/ggWGBS/data/bismark_hg38/{s}_{r}.cpg.methylc\"\n",
    "        mg_fp = f\"/scratch/wzhang/projects/ggWGBS/methylgrapher_no_trim/{s}/{r}/mq10.graph.CG.methyl\"\n",
    "            \n",
    "        methylgrapher_cytosine_data[s][r] = methyl_reader(mg_fp)\n",
    "        bismark_cpg_data[s][r] = methylc_reader(bs_fp)\n",
    "        \n",
    "        \n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "196a85a4baf02a8c",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [
    "# Figure 2A and 2B\n",
    "\n",
    "\n",
    "fig2AB, axes2AB = plt.subplots(1, 2, figsize=(10, 4))\n",
    "\n",
    "\n",
    "slurm_str = \"\"\"\n",
    "\n",
    "\n",
    "400            1.sbatch         20        10G   22:39:40   01:07:59                                   0:0 /scratch/wzhang/pro+        guinness\n",
    "400.batch         batch         20              22:39:40   01:07:59    199140K    932700K                                             guinness\n",
    "402            1.sbatch         20        10G   22:38:00   01:07:54                                   0:0 /scratch/wzhang/pro+        guinness\n",
    "402.batch         batch         20              22:38:00   01:07:54    195192K    929660K                                             guinness\n",
    "403            1.sbatch         20        10G   22:38:40   01:07:56                                   0:0 /scratch/wzhang/pro+        guinness\n",
    "403.batch         batch         20              22:38:40   01:07:56    199800K    932888K                                             guinness\n",
    "\n",
    "\n",
    "404            2.sbatch         20       150G   14:15:00   00:42:45                                   0:0 /scratch/wzhang/pro+          maotai\n",
    "404.batch         batch         20              14:15:00   00:42:45 142916088K 142921640K                                               maotai\n",
    "405            2.sbatch         20       150G   12:50:20   00:38:31                                   0:0 /scratch/wzhang/pro+          maotai\n",
    "405.batch         batch         20              12:50:20   00:38:31 142916396K 142921640K                                               maotai\n",
    "406            2.sbatch         20       150G   12:47:00   00:38:21                                   0:0 /scratch/wzhang/pro+          maotai\n",
    "406.batch         batch         20              12:47:00   00:38:21 142916688K 142921644K                                               maotai\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "420          3.bismark+         20       300G 6-02:55:00   07:20:45                                   0:0 /scratch/wzhang/pro+          maotai\n",
    "420.batch         batch         20            6-02:55:00   07:20:45 274326708K 294423504K                                               maotai\n",
    "421          3.bismark+         20       300G 6-06:15:00   07:30:45                                   0:0 /scratch/wzhang/pro+          maotai\n",
    "421.batch         batch         20            6-06:15:00   07:30:45 292913100K 312429924K                                               maotai\n",
    "422          3.bismark+         20       300G 6-00:03:00   07:12:09                                   0:0 /scratch/wzhang/pro+          maotai\n",
    "422.batch         batch         20            6-00:03:00   07:12:09 287908660K 309915044K                                               maotai\n",
    "\n",
    "\n",
    "\n",
    "720          3.biscuit+         20       100G 4-12:51:00   05:26:33                                   0:0 /scratch/wzhang/pro+        guinness\n",
    "720.batch         batch         20            4-12:51:00   05:26:33  38111276K  42363188K                                             guinness\n",
    "741          3.biscuit+         20       100G 4-12:41:00   05:26:03                                   0:0 /scratch/wzhang/pro+        guinness \n",
    "741.batch         batch         20            4-12:41:00   05:26:03  38102880K  42428728K                                             guinness\n",
    "742          3.biscuit+         20       100G 4-11:59:40   05:23:59                                   0:0 /scratch/wzhang/pro+        guinness \n",
    "742.batch         batch         20            4-11:59:40   05:23:59  38090708K  42690832K                                             guinness\n",
    "\n",
    "700          3.bwameth+         20       100G 5-02:46:00   06:08:18                                   0:0 /scratch/wzhang/pro+        guinness\n",
    "700.batch         batch         20            5-02:46:00   06:08:18  36993548K  41704288K                                             guinness\n",
    "707          3.bwameth+         20       100G 5-02:09:00   06:06:27                                   0:0 /scratch/wzhang/pro+        guinness\n",
    "707.batch         batch         20            5-02:09:00   06:06:27  37025940K  41769848K                                             guinness\n",
    "708          3.bwameth+         20       100G 5-01:27:00   06:04:21                                   0:0 /scratch/wzhang/pro+        guinness\n",
    "708.batch         batch         20            5-01:27:00   06:04:21  37003088K  41573232K                                             guinness\n",
    "\n",
    "\n",
    "706          3.gembs.s+         20       100G 1-12:51:40   01:50:35                                   0:0 /scratch/wzhang/pro+          maotai\n",
    "706.batch         batch         20            1-12:51:40   01:50:35  65610280K  70912712K                                               maotai\n",
    "716          3.gembs.s+         20       100G 1-12:45:40   01:50:17                                   0:0 /scratch/wzhang/pro+          maotai\n",
    "716.batch         batch         20            1-12:45:40   01:50:17  65232744K  70540824K                                               maotai\n",
    "714          3.gembs.s+         20       100G 1-12:57:40   01:50:53                                   0:0 /scratch/wzhang/pro+          maotai\n",
    "714.batch         batch         20            1-12:57:40   01:50:53  65880844K  71135264K                                               maotai\n",
    "\n",
    "\n",
    "\n",
    "727          3.methylg+         20       100G 4-11:00:40   05:21:02                                   0:0 /scratch/wzhang/pro+        tsingtao\n",
    "727.batch         batch         20            4-11:00:40   05:21:02  81600544K 134626188K                                             tsingtao\n",
    "728          3.methylg+         20       100G 4-10:34:20   05:19:43                                   0:0 /scratch/wzhang/pro+        tsingtao\n",
    "728.batch         batch         20            4-10:34:20   05:19:43  81634320K 135462160K                                             tsingtao\n",
    "729          3.methylg+         20       100G 4-10:35:20   05:19:46                                   0:0 /scratch/wzhang/pro+        tsingtao\n",
    "729.batch         batch         20            4-10:35:20   05:19:46  81565756K 134641544K                                             tsingtao\n",
    "\n",
    "\n",
    "\n",
    "410          4.bismark+         20       100G 1-06:59:40   01:32:59                                   0:0 /scratch/wzhang/pro+          maotai\n",
    "410.batch         batch         20            1-06:59:40   01:32:59   2735436K  63275696K                                               maotai\n",
    "411          4.bismark+         20       100G   21:18:20   01:03:55                                   0:0 /scratch/wzhang/pro+          maotai\n",
    "411.batch         batch         20              21:18:20   01:03:55   2810456K  63275700K                                               maotai\n",
    "412          4.bismark+         20       100G   21:08:00   01:03:24                                   0:0 /scratch/wzhang/pro+          maotai\n",
    "412.batch         batch         20              21:08:00   01:03:24   2859192K  63275636K                                               maotai\n",
    "\n",
    "\n",
    "\n",
    "701          4.biscuit+         20       100G 1-07:00:20   01:33:01                                   0:0 /scratch/wzhang/pro+          maotai\n",
    "701.batch         batch         20            1-07:00:20   01:33:01   6537504K   8362704K                                               maotai\n",
    "704          4.biscuit+         20       100G 1-09:50:00   01:41:30                                   0:0 /scratch/wzhang/pro+          maotai\n",
    "704.batch         batch         20            1-09:50:00   01:41:30   4570948K   6018140K                                               maotai\n",
    "718          4.biscuit+         20       100G 1-09:33:20   01:40:40                                   0:0 /scratch/wzhang/pro+          maotai\n",
    "718.batch         batch         20            1-09:33:20   01:40:40   1226048K   1957180K                                               maotai\n",
    "\n",
    "\n",
    "709          4.bwameth+         20       100G   01:45:40   00:05:17                                   0:0 /scratch/wzhang/pro+          maotai\n",
    "709.batch         batch         20              01:45:40   00:05:17    823572K   2100924K                                               maotai\n",
    "710          4.bwameth+         20       100G   01:43:20   00:05:10                                   0:0 /scratch/wzhang/pro+          maotai\n",
    "710.batch         batch         20              01:43:20   00:05:10    820912K   2103784K                                               maotai\n",
    "711          4.bwameth+         20       100G   01:45:20   00:05:16                                   0:0 /scratch/wzhang/pro+          maotai\n",
    "711.batch         batch         20              01:45:20   00:05:16    821544K   2100924K                                               maotai\n",
    "\n",
    "\n",
    "712          4.gembs.s+         20       100G   09:38:20   00:28:55                                   0:0 /scratch/wzhang/pro+          maotai\n",
    "712.batch         batch         20              09:38:20   00:28:55   2200184K  15857552K                                               maotai\n",
    "715          4.gembs.s+         20       100G   09:43:20   00:29:10                                   0:0 /scratch/wzhang/pro+          maotai\n",
    "715.batch         batch         20              09:43:20   00:29:10   2485636K  15817576K                                               maotai\n",
    "717          4.gembs.s+         20       100G   09:47:00   00:29:21                                   0:0 /scratch/wzhang/pro+          maotai\n",
    "717.batch         batch         20              09:47:00   00:29:21   2471200K  15817812K                                               maotai\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "733          4.methylg+         20       100G 2-06:57:00   02:44:51                                   0:0 /scratch/wzhang/pro+       yuengling\n",
    "733.batch         batch         20            2-06:57:00   02:44:51  33076892K  33489496K                                            yuengling\n",
    "735          4.methylg+         20       100G 2-05:49:40   02:41:29                                   0:0 /scratch/wzhang/pro+       yuengling\n",
    "735.batch         batch         20            2-05:49:40   02:41:29  33076808K  33489512K                                            yuengling\n",
    "736          4.methylg+         20       100G 2-06:01:20   02:42:04                                   0:0 /scratch/wzhang/pro+       yuengling\n",
    "736.batch         batch         20            2-06:01:20   02:42:04  33077008K  33490520K                                            yuengling\n",
    "\n",
    "\n",
    "\n",
    "\"\"\".strip().split(\"\\n\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "job_detail = {}\n",
    "for l in slurm_str:\n",
    "    l = l.strip().split()\n",
    "    if len(l) == 0:\n",
    "        continue\n",
    "\n",
    "    job_id = l[0].split(\".\")[0]\n",
    "    job_name = l[1]\n",
    "\n",
    "\n",
    "    # print(job_id, job_name)\n",
    "\n",
    "    if job_id not in job_detail:\n",
    "        job_detail[job_id] = [None, None, None, None]\n",
    "\n",
    "    if job_name != \"batch\":\n",
    "        step = int(job_name.split(\".\")[0])\n",
    "        job_detail[job_id][0] = step\n",
    "\n",
    "        if step >= 3:\n",
    "            pipeline_name = job_name.split(\".\")[1]\n",
    "            pns = None\n",
    "            if pipeline_name.startswith(\"bismark\"):\n",
    "                pns = \"bs\"\n",
    "            elif pipeline_name.startswith(\"biscuit\"):\n",
    "                pns = \"bc\"\n",
    "            elif pipeline_name.startswith(\"bwameth\"):\n",
    "                pns = \"bm\"\n",
    "            elif pipeline_name.startswith(\"gembs\"):\n",
    "                pns = \"gb\"\n",
    "            elif pipeline_name.startswith(\"methylg\"):\n",
    "                pns = \"mg\"\n",
    "            job_detail[job_id][1] = pns\n",
    "            # print(pns)\n",
    "        # Figure out the aligner\n",
    "        # print(job_id, step)\n",
    "\n",
    "    if job_name == \"batch\":\n",
    "        walltime = l[4]\n",
    "        h, m, s = walltime.split(\":\")\n",
    "        walltime = int(h) * 3600 + int(m) * 60 + int(s)\n",
    "        job_detail[job_id][2] = walltime / 3600\n",
    "\n",
    "        rss = int(l[5][:-1]) / 1024 / 1024\n",
    "\n",
    "        job_detail[job_id][3] = rss\n",
    "\n",
    "\n",
    "\n",
    "runtime = {}\n",
    "peak_memory = {}\n",
    "\n",
    "for p in pipeline_name_short:\n",
    "    runtime[p] = [[] for _ in range(5)]\n",
    "    peak_memory[p] = [[] for _ in range(5)]\n",
    "    \n",
    "    runtime[p][0] = [0, 0, 0]\n",
    "    peak_memory[p][0] = [0, 0, 0]\n",
    "\n",
    "for jid in job_detail:\n",
    "    step, pns, rt, pm = job_detail[jid]\n",
    "    # print(jid, step, pipeline, runtime, peak_memory)\n",
    "    \n",
    "    if step in [1, 2]:\n",
    "        for p in pipeline_name_short:\n",
    "            runtime[p][step].append(rt)\n",
    "            peak_memory[p][step].append(pm)\n",
    "            \n",
    "    else:\n",
    "        runtime[pns][step].append(rt)\n",
    "        peak_memory[pns][step].append(pm)\n",
    "\n",
    "\n",
    "for p in pipeline_name_short:\n",
    "    rt_total = numpy.zeros(3)\n",
    "    for i in range(1, 5):\n",
    "        rt_total += runtime[p][i]\n",
    "    \n",
    "    runtime[p][0] = list(rt_total)\n",
    "\n",
    "\n",
    "\n",
    "for p in pipeline_name_short:\n",
    "    plot_x = numpy.arange(5) + 0.15 * (pipeline_name_short.index(p)-2)\n",
    "    plot_y = []\n",
    "    for i in range(5):\n",
    "        print(i, p, numpy.mean(runtime[p][i]))\n",
    "        plot_y.append(numpy.mean(runtime[p][i]))\n",
    "    print()\n",
    "    \n",
    "    axes2AB[0].bar(plot_x, plot_y, width=0.15, label=pipeline_full_name_lookup[p], color=pipeline_color_lookup[p])\n",
    "    axes2AB[0].errorbar(plot_x, plot_y, yerr=[numpy.std(runtime[p][i]) for i in range(5)], fmt='_', color='gray', capsize=2)\n",
    "\n",
    "\n",
    "for p in pipeline_name_short:\n",
    "    plot_x = numpy.arange(4) + 0.15 * (pipeline_name_short.index(p)-2)\n",
    "    plot_y = []\n",
    "    for i in range(1, 5):\n",
    "        print(i, p, numpy.mean(peak_memory[p][i]))\n",
    "        plot_y.append(numpy.mean(peak_memory[p][i]))\n",
    "    print()\n",
    "    \n",
    "    axes2AB[1].bar(plot_x, plot_y, width=0.15, label=pipeline_full_name_lookup[p], color=pipeline_color_lookup[p])\n",
    "    axes2AB[1].errorbar(plot_x, plot_y, yerr=[numpy.std(peak_memory[p][i]) for i in range(1,5)], fmt='_', color='gray', capsize=2)\n",
    "\n",
    "\n",
    "\n",
    "axes2AB[0].set_title(\"Runtime (hour)\")\n",
    "axes2AB[0].set_xticks(numpy.arange(5))\n",
    "axes2AB[0].set_xticklabels([\"Total\", \"Trim\", \"Dedup\", \"Align\", \"Extract\"])\n",
    "\n",
    "axes2AB[1].set_title(\"Peak Memory (GB)\")\n",
    "axes2AB[1].set_xticks(numpy.arange(4))\n",
    "axes2AB[1].set_xticklabels([\"Trim\", \"Dedup\", \"Align\", \"Extract\"])\n",
    "\n",
    "\n",
    "for ax in axes2AB:\n",
    "    ax.set_xticks(numpy.arange(5))\n",
    "    \n",
    "    ax.set_xlabel(\"Steps\")\n",
    "    ax.legend()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "fig2AB.savefig(f\"./figures/2AB.pdf\", bbox_inches='tight')\n",
    "\n",
    "\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "25d3dd542c69b94d",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [
    "# Figure 2C\n",
    "\n",
    "read_fate_bs = {}\n",
    "read_fate_mg = {}\n",
    "\n",
    "\n",
    "# This table is generated from wc -l of fastq files\n",
    "# Therefore, number of read pairs should be divided by 4, number of reads should be divided by 2\n",
    "read_fate_common = \"\"\"HG00621 1       632425760       623983088       593210520\n",
    "HG00621 2       693368864       691954268       653372836\n",
    "HG00741 1       828587944       825623948       785028916\n",
    "HG00741 2       669297276       666649168       638223296\n",
    "HG01952 1       761988944       759292520       740328588\n",
    "HG01952 2       842572440       825532616       784084512\n",
    "HG01978 1       762254308       755087848       734933272\n",
    "HG01978 2       1022644704      1020912880      950967944\n",
    "HG03516 1       1077846324      1074625584      996966416\n",
    "HG03516 2       1131705808      1128505180      1027927044\"\"\".strip().split(\"\\n\")\n",
    "\n",
    "\n",
    "# Common read fate for all methods\n",
    "common_read_fate = {}\n",
    "for rfl in read_fate_common:\n",
    "    l = rfl.strip().split()\n",
    "    s, r, raw_reads, after_trim, after_dedup = l\n",
    "    raw_reads = int(int(raw_reads) / 2)\n",
    "    after_trim = int(int(after_trim) / 2)\n",
    "    after_dedup = int(int(after_dedup) / 2)\n",
    "    \n",
    "    trim = raw_reads - after_trim\n",
    "    dedup = after_trim - after_dedup\n",
    "    \n",
    "    if s not in common_read_fate:\n",
    "        common_read_fate[s] = {}\n",
    "    common_read_fate[s][r] = [raw_reads, trim, dedup, after_dedup]\n",
    "\n",
    "\n",
    "mapping_data = {}\n",
    "\n",
    "# 5 sample, 2 replicate, 5 methods = 50 plots\n",
    "# The mapq distribution plot for aligned reads\n",
    "fig2X, axes2X = plt.subplots(5, 10, figsize=(40, 20))\n",
    "with open(\"./bam_stat_simple/summary.txt\") as fh:\n",
    "    \n",
    "    plot_i = -1\n",
    "    max_mq = 0\n",
    "    for l in fh:\n",
    "        plot_i += 1\n",
    "        \n",
    "        axi = plot_i % 5\n",
    "        axj = plot_i // 5\n",
    "        ax = axes2X[axi][axj]\n",
    "        \n",
    "        l = l.strip().split(\"\\t\", 4)\n",
    "        s, r, psn, mapq_distribution = l\n",
    "        \n",
    "        if s not in mapping_data:\n",
    "            mapping_data[s] = {}\n",
    "        if r not in mapping_data[s]:\n",
    "            mapping_data[s][r] = {}\n",
    "        mapping_data[s][r][psn] = [0, 0]\n",
    "        \n",
    "        # print(s, r, psn, axi, axj)\n",
    "        mapq_distribution = json.loads(mapq_distribution)\n",
    "        \n",
    "        bar = []\n",
    "        for mq in range(255):\n",
    "            mqs = str(mq)\n",
    "            \n",
    "            count = mapq_distribution.get(mqs, 0)\n",
    "            mapping_data[s][r][psn][1] += count\n",
    "            if mq > 10:\n",
    "                mapping_data[s][r][psn][0] += count\n",
    "            \n",
    "            \n",
    "            bar.append(mapq_distribution.get(mqs, 0))\n",
    "            \n",
    "            if count > 0 and mq > max_mq:\n",
    "                max_mq = mq\n",
    "            \n",
    "            \n",
    "        \n",
    "        ax.bar(numpy.arange(255), bar)\n",
    "        ax.set_title(f\"{s}_{r}_{psn}\")\n",
    "\n",
    "    for axi in range(10):\n",
    "        for axj in range(5):\n",
    "            ax = axes2X[axj][axi]\n",
    "            ax.set_xticks(numpy.arange(-10, 70, 10))\n",
    "            ax.set_xlim(-5, 65)\n",
    "            \n",
    "            \n",
    "            \n",
    "        \n",
    "\n",
    "fig2X.savefig(f\"./figures/2X.pdf\", bbox_inches='tight')\n",
    "        \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Old style\n",
    "fig2C, axes2C = plt.subplots(1, 2, sharex=True, figsize=(10, 4))\n",
    "\n",
    "mapping_percentage = {}\n",
    "mapping_percentage_mapq10 = {}\n",
    "for s in samples:\n",
    "    for r in replicates:\n",
    "        r = str(r)\n",
    "        \n",
    "        after_dedup = common_read_fate[s][r][3]\n",
    "        for psn in pipeline_name_short:\n",
    "            \n",
    "            map_count = mapping_data[s][r][psn][1]\n",
    "            map_count_mapq10 = mapping_data[s][r][psn][0]\n",
    "            \n",
    "            map_percentage = map_count / after_dedup * 100\n",
    "            map_percentage_mapq10 = map_count_mapq10 / after_dedup * 100\n",
    "            \n",
    "            if psn not in mapping_percentage:\n",
    "                mapping_percentage[psn] = []\n",
    "                mapping_percentage_mapq10[psn] = []\n",
    "            \n",
    "            mapping_percentage[psn].append(map_percentage)\n",
    "            mapping_percentage_mapq10[psn].append(map_percentage_mapq10)\n",
    "\n",
    "\n",
    "for psn in pipeline_name_short:\n",
    "    plot_x = [pipeline_name_short.index(psn)] * 10\n",
    "    for x in range(10):\n",
    "        plot_x[x] += (x-10) * 0.03\n",
    "    ploy_y1 = mapping_percentage[psn]\n",
    "    ploy_y2 = mapping_percentage_mapq10[psn]\n",
    "\n",
    "    plot_y1_ave = numpy.mean(ploy_y1)\n",
    "    plot_y2_ave = numpy.mean(ploy_y2)\n",
    "    \n",
    "    axes2C[0].scatter(plot_x, ploy_y1, label=pipeline_full_name_lookup[psn], color=pipeline_color_lookup[psn], s=0.5)\n",
    "    axes2C[1].scatter(plot_x, ploy_y2, label=pipeline_full_name_lookup[psn], color=pipeline_color_lookup[psn], s=0.5)\n",
    "\n",
    "    # line half width\n",
    "    axes2C[0].plot([numpy.min(plot_x)-0.1, numpy.max(plot_x)+0.1], [plot_y1_ave, plot_y1_ave], color=pipeline_color_lookup[psn], alpha=0.5, linewidth=1.)\n",
    "    axes2C[1].plot([numpy.min(plot_x)-0.1, numpy.max(plot_x)+0.1], [plot_y2_ave, plot_y2_ave], color=pipeline_color_lookup[psn], alpha=0.5, linewidth=1.)\n",
    "\n",
    "    axes2C[0].text(numpy.mean(plot_x)-0.25, plot_y1_ave-7, f\"{plot_y1_ave:.1f}%\", color=\"black\")\n",
    "    axes2C[1].text(numpy.mean(plot_x)-0.25, plot_y2_ave-7, f\"{plot_y2_ave:.1f}%\", color=\"black\")\n",
    "\n",
    "\n",
    "\n",
    "axes2C[0].set_title(\"Mapping Percentage\")\n",
    "axes2C[1].set_title(\"Mapping Percentage (MapQ > 10)\")\n",
    "for ax in axes2C:\n",
    "    ax.set_xticks(numpy.arange(5))\n",
    "    ax.set_xticklabels([pipeline_full_name_lookup[p] for p in pipeline_name_short])\n",
    "    ax.set_ylabel(\"Mapping Percentage (%)\")\n",
    "    ax.set_ylim(0, 110)\n",
    "    ax.set_yticks(range(0, 110, 20))\n",
    "    \n",
    "    # rotate label\n",
    "    for tick in ax.get_xticklabels():\n",
    "        tick.set_rotation(90)\n",
    "        tick.set_fontsize(8)\n",
    "    # ax.legend()\n",
    "            \n",
    "\n",
    "\n",
    "\n",
    "fig2C.savefig(f\"./figures/2C.pdf\", bbox_inches='tight')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "fig2C, axes2C = plt.subplots(1, 2, sharex=True, figsize=(10, 4))\n",
    "\n",
    "for psn in pipeline_name_short:\n",
    "    plot_x = [pipeline_name_short.index(psn)] * 10\n",
    "    ploy_y1 = mapping_percentage[psn]\n",
    "    ploy_y2 = mapping_percentage_mapq10[psn]\n",
    "\n",
    "    plot_y1_ave = numpy.mean(ploy_y1)\n",
    "    plot_y2_ave = numpy.mean(ploy_y2)\n",
    "\n",
    "    # axes2C[0].boxplot([pipeline_name_short.index(psn)], [ploy_y1], widths=0.03, patch_artist=True, boxprops=dict(facecolor=pipeline_color_lookup[psn]))\n",
    "    # axes2C[1].box(ploy_y1, color=pipeline_color_lookup[psn], alpha=0.5, linewidth=1.)\n",
    "\n",
    "    # Plot y1 in seaborn violin plot\n",
    "    seaborn.violinplot(x=plot_x, y=ploy_y1, ax=axes2C[0], color=pipeline_color_lookup[psn], linewidth=1., width=0.6)\n",
    "    seaborn.violinplot(x=plot_x, y=ploy_y2, ax=axes2C[1], color=pipeline_color_lookup[psn], linewidth=1., width=0.6)\n",
    "\n",
    "\n",
    "    y1_pos = numpy.min(ploy_y1) - 12\n",
    "    y2_pos = numpy.min(ploy_y2) - 12\n",
    "\n",
    "    if y1_pos > 60:\n",
    "        y1_pos += 4\n",
    "\n",
    "    if y2_pos > 60:\n",
    "        y2_pos += 4\n",
    "\n",
    "\n",
    "    axes2C[0].text(numpy.mean(plot_x)-0.25, y1_pos, f\"{plot_y1_ave:.1f}%\", color=\"black\")\n",
    "    axes2C[1].text(numpy.mean(plot_x)-0.25, y2_pos, f\"{plot_y2_ave:.1f}%\", color=\"black\")\n",
    "\n",
    "axes2C[0].set_title(\"Mapping Percentage\")\n",
    "axes2C[1].set_title(\"Mapping Percentage (MapQ > 10)\")\n",
    "for ax in axes2C:\n",
    "    ax.set_xticks(numpy.arange(5))\n",
    "    ax.set_xticklabels([pipeline_full_name_lookup[p] for p in pipeline_name_short])\n",
    "    ax.set_ylabel(\"Mapping Percentage (%)\")\n",
    "    ax.set_ylim(0, 110)\n",
    "    ax.set_yticks(range(0, 110, 20))\n",
    "\n",
    "    # rotate label\n",
    "    for tick in ax.get_xticklabels():\n",
    "        tick.set_rotation(90)\n",
    "        tick.set_fontsize(8)\n",
    "    # ax.legend()\n",
    "\n",
    "\n",
    "fig2C.savefig(f\"./figures/2C.pdf\", bbox_inches='tight')\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "545f1697095f8915",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [
    "# get methylation data\n",
    "# methylation distribution genome wide\n",
    "# CpG count\n",
    "# CpG coverage\n",
    "# CpG methylation rate\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def general_table_opener(fp):\n",
    "    fh = None\n",
    "    if fp.lower().endswith(\".gz\") or fp.lower().endswith(\".gzip\"):\n",
    "        fh = gzip.open(fp, \"rt\")\n",
    "    else:\n",
    "        fh = open(fp, \"r\")\n",
    "\n",
    "    lc = 0\n",
    "    for l in fh:\n",
    "        yield l.strip()\n",
    "        lc += 1\n",
    "        if lc > 1e50:\n",
    "            break\n",
    "\n",
    "    fh.close()\n",
    "\n",
    "\n",
    "def methylc_reader(fp):\n",
    "    res = {}\n",
    "    for l in general_table_opener(fp):\n",
    "        l = l.split(\"\\t\")\n",
    "        chrom, start, end, context, ml, strand, cov = l\n",
    "        start = int(start)\n",
    "        ml = float(ml)\n",
    "        cov = int(cov)\n",
    "        met = int(ml * cov)\n",
    "        if chrom not in res:\n",
    "            res[chrom] = {}\n",
    "\n",
    "        res[chrom][start] = [met, cov]\n",
    "\n",
    "    return res\n",
    "\n",
    "def biscuit_reader(fp):\n",
    "    res = {}\n",
    "    for l in general_table_opener(fp):\n",
    "        l = l.split(\"\\t\")\n",
    "        chrom, start, end, ml, cov, other = l\n",
    "        start = int(start)\n",
    "        ml = float(ml)\n",
    "        cov = int(cov)\n",
    "        met = int(ml * cov)\n",
    "        if chrom not in res:\n",
    "            res[chrom] = {}\n",
    "\n",
    "        res[chrom][start] = [met, cov]\n",
    "\n",
    "    return res\n",
    "\n",
    "def bwameth_reader(fp):\n",
    "    res = {}\n",
    "    for l in general_table_opener(fp):\n",
    "        l = l.split(\"\\t\")\n",
    "        if len(l) != 6:\n",
    "            continue\n",
    "\n",
    "        chrom, start, end, ml, met, unmet = l\n",
    "        start = int(start)\n",
    "        met = int(met)\n",
    "        unmet = int(unmet)\n",
    "        cov = met + unmet\n",
    "\n",
    "        if chrom not in res:\n",
    "            res[chrom] = {}\n",
    "\n",
    "        res[chrom][start] = [met, cov]\n",
    "\n",
    "    return res\n",
    "\n",
    "\n",
    "\n",
    "def methylgrapher_reader(fp):\n",
    "    res_ml = []\n",
    "    res_cov = []\n",
    "    for l in general_table_opener(fp):\n",
    "        cpgid, met, cov = l.split(\"\\t\")\n",
    "        met = int(met)\n",
    "        cov = int(cov)\n",
    "\n",
    "        res_ml.append(met / cov)\n",
    "        res_cov.append(cov)\n",
    "\n",
    "    res_cov = cov2freq(res_cov)\n",
    "    # not including 5\n",
    "    cpg_count_greater_than_cov5 = sum(res_cov[5:])\n",
    "    return res_ml, res_cov, cpg_count_greater_than_cov5\n",
    "\n",
    "\n",
    "def cov2freq(cov, max_cov=30):\n",
    "    res = [0] * max_cov\n",
    "    for c in cov:\n",
    "        if c > 30:\n",
    "            c = 30\n",
    "        res[c-1] += 1\n",
    "    return res\n",
    "\n",
    "\n",
    "def linear_data_to_ml_cov(data):\n",
    "    res_ml = []\n",
    "    res_cov = []\n",
    "    for chrom in data:\n",
    "        for start in data[chrom]:\n",
    "            met, c = data[chrom][start]\n",
    "            res_ml.append(met/c)\n",
    "\n",
    "            res_cov.append(c)\n",
    "\n",
    "    res_cov = cov2freq(res_cov)\n",
    "\n",
    "    # not including 5\n",
    "    cpg_count_greater_than_cov5 = sum(res_cov[5:])\n",
    "    return res_ml, res_cov, cpg_count_greater_than_cov5\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "fig3ab, axes3ab = plt.subplots(1, 2, figsize=(12, 5))\n",
    "fig3a, axes3a = plt.subplots(2, 5, figsize=(22, 8), sharex=True)\n",
    "fig3b, axes3b = plt.subplots(2, 5, figsize=(22, 8), sharex=True)\n",
    "\n",
    "\n",
    "\n",
    "for s in samples:\n",
    "    for r in replicates:\n",
    "        # print(s, r)\n",
    "\n",
    "        si = samples.index(s)\n",
    "        ri = replicates.index(r)\n",
    "        ax3a = axes3a[ri, si]\n",
    "        ax3b = axes3b[ri, si]\n",
    "\n",
    "        bismark_out_fp = f\"/scratch/wzhang/projects/ggWGBS/bismark_hg38_default/{s}/{r}/track.cpg.methylc.gz\"\n",
    "        biscuit_out_fp = f\"/scratch/wzhang/projects/ggWGBS/biscuit_run/{s}/{r}/my_methylation_data.merged.bed.gz\"\n",
    "        bwamem_out_fp = f\"/scratch/wzhang/projects/ggWGBS/bwameth/{s}/{r}/methylation.cpg.bedGraph\"\n",
    "        gembs_out_fp = f\"/scratch/wzhang/projects/ggWGBS/gembs_run/run_merged/8tracks/{s}_{r}/track.cpg.methylc.gz\"\n",
    "        methylgrapher_out_fp = f\"/scratch/wzhang/projects/ggWGBS/methylgrapher_no_trim/{s}/{r}/mq10.graph.CG.merged.tsv\"\n",
    "\n",
    "\n",
    "        bismark_ml = []\n",
    "        biscuit_ml = []\n",
    "        bwamem_ml = []\n",
    "        gembs_ml = []\n",
    "        methylgrapher_ml = []\n",
    "\n",
    "        bismark_cov = []\n",
    "        biscuit_cov = []\n",
    "        bwamem_cov = []\n",
    "        gembs_cov = []\n",
    "        methylgrapher_cov = []\n",
    "\n",
    "\n",
    "        bismark_cpg_data = methylc_reader(bismark_out_fp)\n",
    "        bismark_ml, bismark_cov, bs_cpg_count_cov_greater_than_5 = linear_data_to_ml_cov(bismark_cpg_data)\n",
    "\n",
    "        biscuit_cpg_data = biscuit_reader(biscuit_out_fp)\n",
    "        biscuit_ml, biscuit_cov, bc_cpg_count_cov_greater_than_5 = linear_data_to_ml_cov(biscuit_cpg_data)\n",
    "\n",
    "        bwamem_cpg_data = bwameth_reader(bwamem_out_fp)\n",
    "        bwamem_ml, bwamem_cov, bm_cpg_count_cov_greater_than_5 = linear_data_to_ml_cov(bwamem_cpg_data)\n",
    "\n",
    "        gembs_cpg_data = methylc_reader(gembs_out_fp)\n",
    "        gembs_ml, gembs_cov, gb_cpg_count_cov_greater_than_5 = linear_data_to_ml_cov(gembs_cpg_data)\n",
    "\n",
    "        methylgrapher_ml, methylgrapher_cov, mg_cpg_count_cov_greater_than_5 = methylgrapher_reader(methylgrapher_out_fp)\n",
    "\n",
    "        print(s, r, \"Cov5\", bs_cpg_count_cov_greater_than_5, bc_cpg_count_cov_greater_than_5, bm_cpg_count_cov_greater_than_5, gb_cpg_count_cov_greater_than_5, mg_cpg_count_cov_greater_than_5)\n",
    "        print(s, r, \"All\", len(bismark_ml), len(biscuit_ml), len(bwamem_ml), len(gembs_ml), len(methylgrapher_ml))\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        x = [i+1 for i in range(30)]\n",
    "        for ps in pipeline_name_short:\n",
    "\n",
    "            if ps == \"bs\":\n",
    "                ml = bismark_ml\n",
    "                cov = bismark_cov\n",
    "\n",
    "            if ps == \"bc\":\n",
    "                ml = biscuit_ml\n",
    "                cov = biscuit_cov\n",
    "\n",
    "            if ps == \"bm\":\n",
    "                ml = bwamem_ml\n",
    "                cov = bwamem_cov\n",
    "\n",
    "            if ps == \"gb\":\n",
    "                ml = gembs_ml\n",
    "                cov = gembs_cov\n",
    "\n",
    "            if ps == \"mg\":\n",
    "                ml = methylgrapher_ml\n",
    "                cov = methylgrapher_cov\n",
    "\n",
    "\n",
    "            cov = numpy.array(cov)\n",
    "            cov = cov / numpy.sum(cov)\n",
    "\n",
    "            pipeline_name_full = pipeline_full_name_lookup[ps]\n",
    "            pipeline_color = pipeline_color_lookup[ps]\n",
    "\n",
    "            seaborn.kdeplot(ml, ax=ax3a, label=pipeline_name_full, color=pipeline_color, linestyle=\"-.\", alpha=0.9)\n",
    "            ax3b.plot(x, cov, label=pipeline_name_full, color=pipeline_color, linestyle=\"-.\", alpha=0.9)\n",
    "\n",
    "            if s == \"HG00621\" and r == 1:\n",
    "                seaborn.kdeplot(ml, ax=axes3ab[0], label=pipeline_name_full, color=pipeline_color, linestyle=\"-.\", alpha=0.9)\n",
    "                axes3ab[1].plot(x, cov, label=pipeline_name_full, color=pipeline_color, linestyle=\"-.\", alpha=0.9)\n",
    "\n",
    "\n",
    "                axes3ab[0].set_title(\"CpG Methylation Level Distribution\")\n",
    "                axes3ab[0].set_xlabel(\"Methylation Level\")\n",
    "                axes3ab[0].set_ylabel(\"Density\")\n",
    "                axes3ab[0].legend()\n",
    "\n",
    "                axes3ab[1].set_title(\"Coverage Distribution\")\n",
    "                axes3ab[1].set_xlabel(\"Coverage\")\n",
    "                axes3ab[1].set_ylabel(\"Frequency\")\n",
    "                axes3ab[1].legend()\n",
    "                fig3ab.savefig(\"./figures/3AB.pdf\", dpi=300, bbox_inches=\"tight\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        ax3a.set_xlabel(\"Methylation Level\")\n",
    "        ax3a.set_ylabel(\"Density\")\n",
    "\n",
    "        ax3b.set_xlabel(\"Coverage\")\n",
    "        ax3b.set_ylabel(\"Frequency\")\n",
    "\n",
    "\n",
    "        for ax in [ax3a, ax3b]:\n",
    "            ax.set_title(f\"{s}_{r}\")\n",
    "            ax.legend()\n",
    "\n",
    "\n",
    "    plt.tight_layout()\n",
    "    fig3a.savefig(\"./figures/3A.pdf\", dpi=300, bbox_inches=\"tight\")\n",
    "    fig3b.savefig(\"./figures/3B.pdf\", dpi=300, bbox_inches=\"tight\")\n",
    "\n",
    "\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "e7e2f92d508097c0",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "cpg_count_data_str = f\"\"\"\n",
    "HG00621 1 Cov5 11741845 18071926 18794926 5669852 17858303\n",
    "HG00621 1 All 27163995 27601865 27569426 26182392 28947783\n",
    "HG00621 2 Cov5 15910117 20606666 21343261 11662920 20619053\n",
    "HG00621 2 All 27537025 27728950 27691525 26983109 29179700\n",
    "HG00741 1 Cov5 21660818 24423465 24844556 18058897 24445013\n",
    "HG00741 1 All 27898978 27824855 27770452 27335026 29387774\n",
    "HG00741 2 Cov5 18650025 22053575 22610971 14855493 21929021\n",
    "HG00741 2 All 27768822 27753136 27706786 27194867 29233908\n",
    "HG01952 1 Cov5 14305275 21318588 22966103 13016591 22194644\n",
    "HG01952 1 All 27253305 27698904 27735793 27039254 29330601\n",
    "HG01952 2 Cov5 19498287 23722900 24479310 17068924 24098535\n",
    "HG01952 2 All 27764613 27858866 27820443 27316446 29451874\n",
    "HG01978 1 Cov5 12582107 20737205 22464528 10059453 21538959\n",
    "HG01978 1 All 26965541 27582194 27614979 26748735 29141587\n",
    "HG01978 2 Cov5 22278028 25280946 25881356 20306613 25859772\n",
    "HG01978 2 All 27820561 27817490 27776046 27385860 29452783\n",
    "HG03516 1 Cov5 23280314 25681414 26062664 20189481 26079132\n",
    "HG03516 1 All 27907806 27834517 27778623 27396935 29575158\n",
    "HG03516 2 Cov5 22841755 25624116 26106537 20294108 26153204\n",
    "HG03516 2 All 27842437 27821775 27775678 27385403 29589049\n",
    "\"\"\".strip().split(\"\\n\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "hg38_cpg_count = 29162231\n",
    "graph_cpg_count = 35769868\n",
    "\n",
    "\n",
    "\n",
    "cpg_count_data = {}\n",
    "for s in samples:\n",
    "    cpg_count_data[s] = {}\n",
    "    for r in replicates:\n",
    "        cpg_count_data[s][r] = {}\n",
    "        for p in pipeline_name_short:\n",
    "            cpg_count_data[s][r][p] = [0, 0]\n",
    "\n",
    "\n",
    "for l in cpg_count_data_str:\n",
    "    l = l.strip().split()\n",
    "    for i in range(3, 8):\n",
    "        l[i] = int(l[i])\n",
    "\n",
    "    s, r, t, bs, bc, bm, gb, mg = l\n",
    "    r = int(r)\n",
    "\n",
    "    list_index = 0\n",
    "    if t == \"All\":\n",
    "        list_index = 1\n",
    "\n",
    "    cpg_count_data[s][r][\"bs\"][list_index] = bs\n",
    "    cpg_count_data[s][r][\"bc\"][list_index] = bc\n",
    "    cpg_count_data[s][r][\"bm\"][list_index] = bm\n",
    "    cpg_count_data[s][r][\"gb\"][list_index] = gb\n",
    "    cpg_count_data[s][r][\"mg\"][list_index] = mg\n",
    "\n",
    "    # print(s, r, t, bs, bc, bm, gb, mg)\n",
    "\n",
    "\n",
    "\n",
    "fig3C, axes3C = plt.subplots(2, 1, figsize=(10, 8))\n",
    "\n",
    "\n",
    "ax3C = axes3C[0]\n",
    "for p in pipeline_name_short:\n",
    "    x0_offset = 0.15 * (pipeline_name_short.index(p) - 2)\n",
    "\n",
    "\n",
    "    x = []\n",
    "    y1 = []\n",
    "    y2 = []\n",
    "\n",
    "    x0 = 0\n",
    "    for s in samples:\n",
    "        for r in replicates:\n",
    "            y1.append(cpg_count_data[s][r][p][0])\n",
    "            y2.append(cpg_count_data[s][r][p][1])\n",
    "\n",
    "            x.append(x0 + x0_offset)\n",
    "            x0 += 1\n",
    "    \n",
    "    # print(min(y2), max(y2))\n",
    "    ax3C.bar(x, y1, width=0.15, label=f\"{pipeline_full_name_lookup[p]}\", color=pipeline_color_lookup[p])\n",
    "    ax3C.bar(x, y2, width=0.15, color=pipeline_color_lookup[p], alpha=0.5)\n",
    "\n",
    "\n",
    "\n",
    "for x0 in range(10):\n",
    "    ax3C.plot([x0-0.4, x0+0.2], [hg38_cpg_count, hg38_cpg_count],   color=\"black\", alpha=0.5)\n",
    "    ax3C.plot([x0+0.2, x0+0.35], [graph_cpg_count, graph_cpg_count], color=\"black\", alpha=0.5)\n",
    "\n",
    "\n",
    "\n",
    "ax3C.text(-0.7, hg38_cpg_count + 1e6, f\"hg38: {hg38_cpg_count/1e6:.1f}M\", color=\"black\")\n",
    "ax3C.text(-0.7, graph_cpg_count + 1e6, f\"HPRC: {graph_cpg_count/1e6:.1f}M\", color=\"black\")\n",
    "\n",
    "ax3C.set_xticks(numpy.arange(10))\n",
    "ax3C.set_xticklabels([f\"{r}\" for s in samples for r in replicates])\n",
    "ax3C.set_title(\"CpG Count\")\n",
    "ax3C.legend(loc='upper left', bbox_to_anchor=(1, 1))\n",
    "ax3C.set_ylim(0, 3.9e7)\n",
    "\n",
    "ax3C.set_yticks([0, 1e7, 2e7, 3e7])\n",
    "ax3C.set_yticklabels([\"\", \"10M\", \"20M\", \"30M\"])\n",
    "# axes3C.set_ylabel(\"Count\")\n",
    "\n",
    "\n",
    "\n",
    "axes3C[1].bar([0], [0], width=0.15, label=f\"  Cov>5\", color=\"gray\")\n",
    "axes3C[1].bar([0], [0], width=0.15, label=f\"0<Cov<=5\", color=\"gray\", alpha=0.5)\n",
    "axes3C[1].legend(loc='upper left', bbox_to_anchor=(1, 1))\n",
    "\n",
    "\n",
    "fig3C.savefig(\"./figures/3C.pdf\", dpi=300, bbox_inches=\"tight\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "9063cdbe2078fce9",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [
    "# Figure 4 Helper Functions\n",
    "\n",
    "\n",
    "def linear_regression(x, y):\n",
    "    x = np.array(x).reshape((-1, 1))\n",
    "    y = np.array(y)\n",
    "\n",
    "    model = LinearRegression()\n",
    "    model.fit(x, y)\n",
    "\n",
    "    k = model.coef_[0]\n",
    "    b = model.intercept_\n",
    "\n",
    "    r_sq = model.score(x, y)\n",
    "\n",
    "    return k, b, r_sq\n",
    "\n",
    "def pearson(x, y):\n",
    "    x = np.array(x)\n",
    "    y = np.array(y)\n",
    "    r = scipy.stats.pearsonr(x, y)[0]\n",
    "    return r\n",
    "\n",
    "\n",
    "def array_fix(hmd):\n",
    "    dim0 = hmd.shape[0]\n",
    "\n",
    "    newhmd = np.zeros(hmd.shape)\n",
    "    for xo in range(dim0):\n",
    "        for yo in range(dim0):\n",
    "            xn = dim0-1 - yo\n",
    "            yn = xo\n",
    "            newhmd[xn, yn] = hmd[xo, yo]\n",
    "\n",
    "    return newhmd\n",
    "\n",
    "\n",
    "def density2d(x, y, bin=40, log=None):\n",
    "    assert len(x) == len(y)\n",
    "\n",
    "    shape = (bin+1, bin+1)\n",
    "    step = 1 / bin\n",
    "    res = numpy.ones(shape)\n",
    "\n",
    "    total = 0\n",
    "    for i in range(len(x)):\n",
    "        x0 = x[i]\n",
    "        y0 = y[i]\n",
    "\n",
    "        # yi = bin - int(x0 / step)\n",
    "        # xi = int(y0 / step)\n",
    "        xi = int(x0 / step)\n",
    "        yi = int(y0 / step)\n",
    "\n",
    "        # print(x0, xi, y0, yi, step)\n",
    "\n",
    "        res[xi, yi] += 1\n",
    "\n",
    "    mm = numpy.max(res)\n",
    "    total = numpy.sum(res)\n",
    "\n",
    "    for a in range(bin+1):\n",
    "        for b in range(bin+1):\n",
    "            pass\n",
    "            # print(res[a, b])\n",
    "            # res[a, b] = math.log2(res[a, b]/total)\n",
    "            res[a, b] = res[a, b] / mm\n",
    "            if log != None:\n",
    "                res[a, b] = math.log2(res[a, b])\n",
    "            # print(res[a, b])\n",
    "\n",
    "\n",
    "    return array_fix(res)\n",
    "\n",
    "\n",
    "\n",
    "def density2dInt(x, y, log=None):\n",
    "    assert len(x) == len(y)\n",
    "\n",
    "    lowest = 0\n",
    "    highest = max(x+y)\n",
    "\n",
    "    shape = (highest+1, highest+1)\n",
    "    step = 1\n",
    "    res = numpy.ones(shape)\n",
    "\n",
    "    total = 0\n",
    "    for i in range(len(x)):\n",
    "        x0 = x[i]\n",
    "        y0 = y[i]\n",
    "\n",
    "        # yi = bin - int(x0 / step)\n",
    "        # xi = int(y0 / step)\n",
    "        xi = int(x0 / step)\n",
    "        yi = int(y0 / step)\n",
    "\n",
    "        # print(x0, xi, y0, yi, step)\n",
    "\n",
    "        res[xi, yi] += 1\n",
    "\n",
    "    mm = numpy.max(res)\n",
    "    total = numpy.sum(res)\n",
    "\n",
    "    for a in range(highest+1):\n",
    "        for b in range(highest+1):\n",
    "            pass\n",
    "            # print(res[a, b])\n",
    "            # res[a, b] = math.log2(res[a, b]/total)\n",
    "            res[a, b] = res[a, b] / mm\n",
    "            if log != None:\n",
    "                res[a, b] = math.log2(res[a, b])\n",
    "            # print(res[a, b])\n",
    "\n",
    "\n",
    "    return array_fix(res)\n",
    "\n",
    "\n",
    "def diff_bar_graph_data(delta):\n",
    "\n",
    "    delta_x = []\n",
    "    delta_y = []\n",
    "\n",
    "\n",
    "    diff = [0.1, 0.2, 0.3, 0.4, 0.5]\n",
    "    for i in range(1, 10):\n",
    "        diff_range_center = i / 10 - 0.5\n",
    "        delta_x.append(diff_range_center)\n",
    "        delta_y.append(0)\n",
    "\n",
    "\n",
    "    for d in delta:\n",
    "        if d < -0.45:\n",
    "            d = -0.45\n",
    "        if d > 0.44:\n",
    "            d = 0.44\n",
    "\n",
    "        for idx, dc in enumerate(delta_x):\n",
    "            if abs(dc - d) < 0.05:\n",
    "                delta_y[idx] += 1\n",
    "\n",
    "    return delta_x, delta_y\n",
    "\n",
    "\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "9e0190a784b422",
   "metadata": {},
   "source": [
    "# Figure S4-S7, and 4A, 4B\n",
    "\n",
    "\n",
    "\n",
    "for pns in pipeline_name_short:\n",
    "    if pns == \"mg\":\n",
    "        continue\n",
    "\n",
    "    print(pns)\n",
    "    fig4A, axes4A = plt.subplots(3, 5, figsize=(25, 18))\n",
    "    plt.subplots_adjust(hspace=0.4)\n",
    "    for s in samples:\n",
    "        print(\"\\t\", s)\n",
    "\n",
    "        ml_delta = []\n",
    "\n",
    "        bs_cpg_cov_r1 = []\n",
    "        bs_cpg_cov_r2 = []\n",
    "\n",
    "        bs_cpg_ml_r1 = []\n",
    "        bs_cpg_ml_r2 = []\n",
    "\n",
    "\n",
    "        #bsd_r1 = bismark_cpg_data[s][replicates[0]]\n",
    "        #bsd_r2 = bismark_cpg_data[s][replicates[1]]\n",
    "\n",
    "        bsd_r1 = methylc_getter(pns, \"cpg\", s, replicates[0])\n",
    "        bsd_r2 = methylc_getter(pns, \"cpg\", s, replicates[1])\n",
    "\n",
    "\n",
    "        for chrom in bsd_r1:\n",
    "            if chrom not in bsd_r2:\n",
    "                continue\n",
    "\n",
    "            for pos in bsd_r1[chrom]:\n",
    "                pos = int(pos)\n",
    "\n",
    "                if pos not in bsd_r2[chrom]:\n",
    "                    continue\n",
    "\n",
    "                bsd01 = bsd_r1[chrom][pos]\n",
    "                bsd02 = bsd_r2[chrom][pos]\n",
    "\n",
    "                cov1 = bsd01[1]\n",
    "                cov2 = bsd02[1]\n",
    "                if cov1 > 30:\n",
    "                    cov1 = 30\n",
    "                if cov2 > 30:\n",
    "                    cov2 = 30\n",
    "                bs_cpg_cov_r1.append(cov1)\n",
    "                bs_cpg_cov_r2.append(cov2)\n",
    "\n",
    "                if bsd01[1] <= min_cov or bsd02[1] <= min_cov:\n",
    "                    continue\n",
    "\n",
    "                ml1 = bsd01[0] / bsd01[1]\n",
    "                ml2 = bsd02[0] / bsd02[1]\n",
    "\n",
    "                diff = ml1 - ml2\n",
    "                ml_delta.append(diff)\n",
    "\n",
    "                bs_cpg_ml_r1.append(ml1)\n",
    "                bs_cpg_ml_r2.append(ml2)\n",
    "\n",
    "\n",
    "        ml_k, ml_b, ml_r2 = linear_regression(bs_cpg_ml_r1, bs_cpg_ml_r2)\n",
    "        ml_r_pearson = pearson(bs_cpg_ml_r1, bs_cpg_ml_r2)\n",
    "\n",
    "        ml_hmd = density2d(bs_cpg_ml_r1, bs_cpg_ml_r2, bin=60)\n",
    "        ml_hmd = gaussian_filter(ml_hmd, sigma=2)\n",
    "\n",
    "\n",
    "        cov_k, cov_b, cov_r2 = linear_regression(bs_cpg_cov_r1, bs_cpg_cov_r2)\n",
    "        cov_r_pearson = pearson(bs_cpg_cov_r1, bs_cpg_cov_r2)\n",
    "\n",
    "        cov_hmd = density2dInt(bs_cpg_cov_r1, bs_cpg_cov_r2)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        dim0 = ml_hmd.shape[0]\n",
    "        bcount = dim0-1\n",
    "        bcount_half = int(bcount/2)\n",
    "\n",
    "\n",
    "\n",
    "        axes_xi = samples.index(s)\n",
    "\n",
    "\n",
    "        ax1 = axes4A[0, axes_xi]\n",
    "        ax2 = axes4A[1, axes_xi]\n",
    "        ax3 = axes4A[2, axes_xi]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        # hmd_gb = gaussian_filter(hmd, sigma=2)\n",
    "        im = ax1.imshow(ml_hmd, cmap=\"Blues\", vmin=0, vmax=0.02) # , vmin=0, vmax=0.02\n",
    "\n",
    "\n",
    "        ax1.set_title(f\"{s}\")\n",
    "        ax1.set_xlabel(\"Methylation (Rep 1)\")\n",
    "        ax1.set_ylabel(\"Methylation (Rep 2)\")\n",
    "        ax1.text(32, 56, f\"y = {ml_k:0.2f}x + {ml_b:0.2f}\\nr = {ml_r_pearson:0.2f}\\nN = {len(bs_cpg_ml_r1)/1e6:0.2f} M\", fontsize=12, color=\"black\", bbox=dict(facecolor='white', alpha=0.0))\n",
    "        ax1.set_xticks([0, 0.25*bcount, 0.5*bcount, 0.75*bcount, bcount], [0, 0.25, 0.5, 0.75, 1])\n",
    "        ax1.set_yticks([bcount, 0.75*bcount, 0.5*bcount, 0.25*bcount, 0], [0, 0.25, 0.5, 0.75, 1])\n",
    "\n",
    "        ax3.imshow(cov_hmd, cmap=\"Blues\", vmin=0, )\n",
    "        ax3.text(16, 28, f\"y = {cov_k:0.2f}x + {cov_b:0.2f}\\nr = {cov_r_pearson:0.2f}\\nN = {len(bs_cpg_cov_r1)/1e6:0.2f} M\", fontsize=12, color=\"black\", bbox=dict(facecolor='white', alpha=0.0))\n",
    "        ax3.set_title(f\"{s}\")\n",
    "\n",
    "        ax3.set_title(f\"Coverage 2D histogram\")\n",
    "        ax3.set_xlabel(\"Coverage (Rep 1)\")\n",
    "        ax3.set_ylabel(\"Coverage (Rep 2)\")\n",
    "        ax3.set_xticks([0, 5, 10, 15, 20, 25, 30], [0, 5, 10, 15, 20, 25, 30])\n",
    "        ax3.set_yticks([0, 5, 10, 15, 20, 25, 30], [30, 25, 20, 15, 10, 5, 0])\n",
    "\n",
    "\n",
    "        # ax1.colorbar(im, fraction=0.02)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        # seaborn.kdeplot(np.array(ml_delta), bw_method=0.1, ax=ax2)\n",
    "        delta_x, delta_y = diff_bar_graph_data(ml_delta)\n",
    "        ax2.bar(delta_x, delta_y, width=0.05)\n",
    "\n",
    "        # Updates\n",
    "        ax2.set_title(f\"Delta Methylation\")\n",
    "        ax2.set_xlabel(\"Methylation Diff\")\n",
    "        ax2.set_ylabel(\"# of CpG (in millions)\")\n",
    "\n",
    "        y0 = max(delta_y) * 1.02\n",
    "        ax2.plot([-0.43, 0.43], [y0, y0], color='#1f76b4',)\n",
    "        ax2.plot([0.15, 0.15], [0, y0], linestyle='dashed', color='#1f76b4',)\n",
    "        ax2.plot([-0.15, -0.15], [0, y0], linestyle='dashed', color='#1f76b4',)\n",
    "\n",
    "        t = [-0.45, -0.35, -0.25, -0.15, -0.05, 0.05, 0.15, 0.25, 0.35, 0.45]\n",
    "        tl = [-1, -0.35, -0.25, -0.15, -0.05, 0.05, 0.15, 0.25, 0.35, 1]\n",
    "        ax2.set_xticks(t, tl, rotation='vertical')\n",
    "\n",
    "        a = sum(delta_y[:3]) / sum(delta_y) * 100\n",
    "        b = sum(delta_y[3:6]) / sum(delta_y) * 100\n",
    "        c = sum(delta_y[6:]) / sum(delta_y) * 100\n",
    "\n",
    "        ax2.text(-0.35, y0*1.01, f\"{a:0.1f} %\", fontsize=10, color=\"#1f76b4\")\n",
    "        ax2.text(-0.05, y0*1.01, f\"{b:0.1f} %\", fontsize=10, color=\"#1f76b4\")\n",
    "        ax2.text( 0.25, y0*1.01, f\"{c:0.1f} %\", fontsize=10, color=\"#1f76b4\")\n",
    "\n",
    "        concordance = b\n",
    "\n",
    "\n",
    "        fig4A.savefig(f\"./figures/S4_8.{pns}.pdf\", bbox_inches=\"tight\")\n",
    "\n",
    "        assert len(bs_cpg_ml_r1) == len(bs_cpg_ml_r2)\n",
    "        print(f\"\\t\\tY= {ml_k:0.4f}X + {ml_b:0.4f}, R^2 = {ml_r2:0.4f}, Pearson = {ml_r_pearson:0.4f}, N={len(bs_cpg_ml_r1)/1e6:0.1f}M, concordance = {concordance:0.2f}%\")\n",
    "        print(f\"\\t\\tY= {cov_k:0.4f}X + {cov_b:0.4f}, R^2 = {cov_r2:0.4f}, Pearson = {cov_r_pearson:0.4f} N={len(bs_cpg_cov_r1)/1e6:0.1f}M\")\n",
    "\n",
    "\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "90c2764eaf0da75a",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [
    "# Figure S8, 4C, 4D\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Make subplot space larger\n",
    "fig4B, axes4B = plt.subplots(3, 5, figsize=(25, 18))\n",
    "plt.subplots_adjust(hspace=0.4)\n",
    "\n",
    "print(\"mg\")\n",
    "for s in samples:\n",
    "    print(\"\\t\"+s)\n",
    "\n",
    "\n",
    "    ml_delta = []\n",
    "    mg_cpg_ml_r1 = []\n",
    "    mg_cpg_ml_r2 = []\n",
    "\n",
    "    mg_cpg_cov_r1 = []\n",
    "    mg_cpg_cov_r2 = []\n",
    "\n",
    "    mgd_r1 = methylgrapher_cpg_data[s][replicates[0]]\n",
    "    mgd_r2 = methylgrapher_cpg_data[s][replicates[1]]\n",
    "\n",
    "    for cpg_id in mgd_r1:\n",
    "        if cpg_id not in mgd_r2:\n",
    "            continue\n",
    "\n",
    "\n",
    "        mgd01 = mgd_r1[cpg_id]\n",
    "        mgd02 = mgd_r2[cpg_id]\n",
    "\n",
    "        cov1 = mgd01[1]\n",
    "        cov2 = mgd02[1]\n",
    "        if cov1 > 30:\n",
    "            cov1 = 30\n",
    "        if cov2 > 30:\n",
    "            cov2 = 30\n",
    "        mg_cpg_cov_r1.append(cov1)\n",
    "        mg_cpg_cov_r2.append(cov2)\n",
    "\n",
    "        if mgd01[1] <= min_cov or mgd02[1] <= min_cov:\n",
    "            continue\n",
    "\n",
    "        ml1 = mgd01[0] / mgd01[1]\n",
    "        ml2 = mgd02[0] / mgd02[1]\n",
    "\n",
    "        diff = ml1 - ml2\n",
    "        ml_delta.append(diff)\n",
    "\n",
    "        mg_cpg_ml_r1.append(ml1)\n",
    "        mg_cpg_ml_r2.append(ml2)\n",
    "\n",
    "\n",
    "\n",
    "    ml_k, ml_b, ml_r2 = linear_regression(mg_cpg_ml_r1, mg_cpg_ml_r2)\n",
    "    ml_r_pearson = pearson(mg_cpg_ml_r1, mg_cpg_ml_r2)\n",
    "\n",
    "    ml_hmd = density2d(mg_cpg_ml_r1, mg_cpg_ml_r2, bin=60)\n",
    "\n",
    "\n",
    "    cov_k, cov_b, cov_r2 = linear_regression(mg_cpg_cov_r1, mg_cpg_cov_r2)\n",
    "    cov_r_pearson = pearson(mg_cpg_cov_r1, mg_cpg_cov_r2)\n",
    "\n",
    "    cov_hmd = density2dInt(mg_cpg_cov_r1, mg_cpg_cov_r2)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    dim0 = ml_hmd.shape[0]\n",
    "    bcount = dim0-1\n",
    "    bcount_half = int(bcount/2)\n",
    "\n",
    "\n",
    "\n",
    "    axes_xi = samples.index(s)\n",
    "\n",
    "\n",
    "    ax1 = axes4B[0, axes_xi]\n",
    "    ax2 = axes4B[1, axes_xi]\n",
    "    ax3 = axes4B[2, axes_xi]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    ml_hmd = gaussian_filter(ml_hmd, sigma=2)\n",
    "    im = ax1.imshow(ml_hmd, cmap=\"Blues\", vmin=0, vmax=0.02) # , vmin=0, vmax=0.02\n",
    "\n",
    "\n",
    "    ax1.set_title(f\"{s}\")\n",
    "    ax1.set_xlabel(\"Methylation (Rep 1)\")\n",
    "    ax1.set_ylabel(\"Methylation (Rep 2)\")\n",
    "    ax1.text(32, 56, f\"y = {ml_k:0.2f}x + {ml_b:0.2f}\\nr = {ml_r_pearson:0.2f}\\nN = {len(mg_cpg_ml_r1)/1e6:0.2f} M\", fontsize=12, color=\"black\", bbox=dict(facecolor='white', alpha=0.0))\n",
    "    ax1.set_xticks([0, 0.25*bcount, 0.5*bcount, 0.75*bcount, bcount], [0, 0.25, 0.5, 0.75, 1])\n",
    "    ax1.set_yticks([bcount, 0.75*bcount, 0.5*bcount, 0.25*bcount, 0], [0, 0.25, 0.5, 0.75, 1])\n",
    "\n",
    "    ax3.imshow(cov_hmd, cmap=\"Blues\", vmin=0, )\n",
    "    ax3.text(16, 28, f\"y = {cov_k:0.2f}x + {cov_b:0.2f}\\nr = {cov_r_pearson:0.2f}\\nN = {len(mg_cpg_cov_r1)/1e6:0.2f} M\", fontsize=12, color=\"black\", bbox=dict(facecolor='white', alpha=0.0))\n",
    "    ax3.set_title(f\"{s}\")\n",
    "\n",
    "    ax3.set_title(f\"Coverage 2D histogram\")\n",
    "    ax3.set_xlabel(\"Coverage (Rep 1)\")\n",
    "    ax3.set_ylabel(\"Coverage (Rep 2)\")\n",
    "    ax3.set_xticks([0, 5, 10, 15, 20, 25, 30], [0, 5, 10, 15, 20, 25, 30])\n",
    "    ax3.set_yticks([0, 5, 10, 15, 20, 25, 30], [30, 25, 20, 15, 10, 5, 0])\n",
    "\n",
    "\n",
    "    # ax1.colorbar(im, fraction=0.02)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    # seaborn.kdeplot(np.array(ml_delta), bw_method=0.1, ax=ax2)\n",
    "    delta_x, delta_y = diff_bar_graph_data(ml_delta)\n",
    "    ax2.bar(delta_x, delta_y, width=0.05)\n",
    "\n",
    "    # Updates\n",
    "    ax2.set_title(f\"Delta Methylation\")\n",
    "    ax2.set_xlabel(\"Methylation Diff\")\n",
    "    ax2.set_ylabel(\"# of CpG (in millions)\")\n",
    "\n",
    "    y0 = max(delta_y) * 1.02\n",
    "    ax2.plot([-0.43, 0.43], [y0, y0], color='#1f76b4',)\n",
    "    ax2.plot([0.15, 0.15], [0, y0], linestyle='dashed', color='#1f76b4',)\n",
    "    ax2.plot([-0.15, -0.15], [0, y0], linestyle='dashed', color='#1f76b4',)\n",
    "\n",
    "    t = [-0.45, -0.35, -0.25, -0.15, -0.05, 0.05, 0.15, 0.25, 0.35, 0.45]\n",
    "    tl = [-1, -0.35, -0.25, -0.15, -0.05, 0.05, 0.15, 0.25, 0.35, 1]\n",
    "    ax2.set_xticks(t, tl, rotation='vertical')\n",
    "\n",
    "    a = sum(delta_y[:3]) / sum(delta_y) * 100\n",
    "    b = sum(delta_y[3:6]) / sum(delta_y) * 100\n",
    "    c = sum(delta_y[6:]) / sum(delta_y) * 100\n",
    "\n",
    "    ax2.text(-0.35, y0*1.01, f\"{a:0.1f} %\", fontsize=10, color=\"#1f76b4\")\n",
    "    ax2.text(-0.05, y0*1.01, f\"{b:0.1f} %\", fontsize=10, color=\"#1f76b4\")\n",
    "    ax2.text( 0.25, y0*1.01, f\"{c:0.1f} %\", fontsize=10, color=\"#1f76b4\")\n",
    "\n",
    "    concordance = b\n",
    "\n",
    "\n",
    "\n",
    "    fig4B.savefig(f\"./figures/S4_8.mg.pdf\", bbox_inches=\"tight\")\n",
    "\n",
    "    assert len(mg_cpg_cov_r1) == len(mg_cpg_cov_r2)\n",
    "    print(f\"\\t\\tY= {ml_k:0.4f}X + {ml_b:0.4f}, R^2 = {ml_r2:0.4f}, Pearson = {ml_r_pearson:0.4f}, N={len(mg_cpg_ml_r1)/1e6:0.1f}M, concordance = {concordance:0.2f}%\")\n",
    "    print(f\"\\t\\tY= {cov_k:0.4f}X + {cov_b:0.4f}, R^2 = {cov_r2:0.4f}, Pearson = {cov_r_pearson:0.4f}, N={len(mg_cpg_cov_r2)/1e6:0.1f}M\")\n",
    "    print()\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "82796ffa3a31df92",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [
    "# Figure S9_13, 4E, 4F\n",
    "\n",
    "\n",
    "log_fh = open(\"figures/S9_13_4E_4F.log\", \"w\")\n",
    "\n",
    "for pns in pipeline_name_short:\n",
    "    if pns == \"mg\":\n",
    "        continue\n",
    "    \n",
    "    print(pns)\n",
    "    fig4C, axes4C = plt.subplots(6, 5, figsize=(25, 40))\n",
    "    plt.subplots_adjust(hspace=0.4)\n",
    "    for s in samples:\n",
    "        for r in replicates:\n",
    "\n",
    "\n",
    "            print(\"\\t\", s, r)\n",
    "\n",
    "            ml_delta = []\n",
    "\n",
    "            bs_cpg_ml = []\n",
    "            mg_cpg_ml = []\n",
    "\n",
    "            bs_cpg_cov = []\n",
    "            mg_cpg_cov = []\n",
    "\n",
    "            # \"bsd\" used to mean bismark data, now it means linear method data\n",
    "            # bsd = bismark_cpg_data[s][r]\n",
    "            bsd = methylc_getter(pns, \"cpg\", s, r)\n",
    "            mg2ld = methylgrapher_lifted_cytosine_data[s][r]\n",
    "\n",
    "            found = 0\n",
    "            for chrom in bsd:\n",
    "                if chrom not in mg2ld:\n",
    "                    continue\n",
    "\n",
    "                for pos in bsd[chrom]:\n",
    "                    pos = int(pos)\n",
    "\n",
    "                    bsd0 = bsd[chrom][pos]\n",
    "                    ml_bs = bsd0[0] / bsd0[1]\n",
    "\n",
    "\n",
    "                    mgd01 = mg2ld[chrom].get(pos, (0, 0))\n",
    "                    mgd02 = mg2ld[chrom].get(pos+1, (0, 0))\n",
    "\n",
    "                    mg_met = mgd01[0] + mgd02[0]\n",
    "                    mg_cov = mgd01[1] + mgd02[1]\n",
    "\n",
    "                    if mg_cov == 0:\n",
    "                        continue\n",
    "\n",
    "                    bs_cov0 = bsd0[1]\n",
    "                    mg_cov0 = mg_cov\n",
    "\n",
    "                    if bs_cov0 > 30:\n",
    "                        bs_cov0 = 30\n",
    "                    if mg_cov0 > 30:\n",
    "                        mg_cov0 = 30\n",
    "\n",
    "                    bs_cpg_cov.append(bs_cov0)\n",
    "                    mg_cpg_cov.append(mg_cov0)\n",
    "\n",
    "\n",
    "                    if bsd0[1] <= min_cov:\n",
    "                        continue\n",
    "\n",
    "                    if mg_cov <= min_cov:\n",
    "                        continue\n",
    "\n",
    "                    ml_mg = mg_met / mg_cov\n",
    "\n",
    "                    # print(ml_bs, f\"{mg_met}/{mg_cov}\", mgd01, mgd02)\n",
    "\n",
    "                    diff = ml_bs - ml_mg\n",
    "                    ml_delta.append(diff)\n",
    "\n",
    "                    bs_cpg_ml.append(ml_bs)\n",
    "                    mg_cpg_ml.append(ml_mg)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "            ml_k, ml_b, ml_r2 = linear_regression(bs_cpg_ml, mg_cpg_ml)\n",
    "            ml_r_pearson = pearson(bs_cpg_ml, mg_cpg_ml)\n",
    "\n",
    "            ml_hmd = density2d(bs_cpg_ml, mg_cpg_ml, bin=60)\n",
    "\n",
    "\n",
    "            cov_k, cov_b, cov_r2 = linear_regression(bs_cpg_cov, mg_cpg_cov)\n",
    "            cov_r_pearson = pearson(bs_cpg_cov, mg_cpg_cov)\n",
    "\n",
    "            cov_hmd = density2dInt(bs_cpg_cov, mg_cpg_cov)\n",
    "\n",
    "\n",
    "\n",
    "            dim0 = ml_hmd.shape[0]\n",
    "            bcount = dim0-1\n",
    "            bcount_half = int(bcount/2)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "            axes_xi = samples.index(s)\n",
    "            axes_yi = replicates.index(r)\n",
    "            if axes_yi == 1:\n",
    "                axes_yi += 2\n",
    "\n",
    "            ax1 = axes4C[axes_yi, axes_xi]\n",
    "            ax2 = axes4C[axes_yi+1, axes_xi]\n",
    "            ax3 = axes4C[axes_yi+2, axes_xi]\n",
    "\n",
    "\n",
    "            ml_hmd = gaussian_filter(ml_hmd, sigma=2)\n",
    "            im = ax1.imshow(ml_hmd, cmap=\"Blues\", vmin=0, vmax=0.02) # , vmin=0, vmax=0.02\n",
    "\n",
    "\n",
    "            ax1.set_title(f\"{s}_{r}\")\n",
    "            ax1.set_xlabel(f\"Methylation ({pipeline_full_name_lookup[pns]})\")\n",
    "            ax1.set_ylabel(\"Methylation (methylGrapher)\")\n",
    "            ax1.text(32, 56, f\"y = {ml_k:0.2f}x + {ml_b:0.2f}\\nr = {ml_r_pearson:0.2f}\\nN = {len(mg_cpg_ml)/1e6:0.2f} M\", fontsize=12, color=\"black\", bbox=dict(facecolor='white', alpha=0.0))\n",
    "            ax1.set_xticks([0, 0.25*bcount, 0.5*bcount, 0.75*bcount, bcount], [0, 0.25, 0.5, 0.75, 1])\n",
    "            ax1.set_yticks([bcount, 0.75*bcount, 0.5*bcount, 0.25*bcount, 0], [0, 0.25, 0.5, 0.75, 1])\n",
    "\n",
    "            ax3.imshow(cov_hmd, cmap=\"Blues\", vmin=0, )\n",
    "            ax3.text(16, 28, f\"y = {cov_k:0.2f}x + {cov_b:0.2f}\\nr = {cov_r_pearson:0.2f}\\nN = {len(mg_cpg_cov)/1e6:0.2f} M\", fontsize=12, color=\"black\", bbox=dict(facecolor='white', alpha=0.0))\n",
    "            ax3.set_title(f\"{s}\")\n",
    "\n",
    "            ax3.set_title(f\"Coverage 2D histogram\")\n",
    "            ax3.set_xlabel(f\"Coverage ({pipeline_full_name_lookup[pns]})\")\n",
    "            ax3.set_ylabel(f\"Coverage (methylGrapher)\")\n",
    "            ax3.set_xticks([0, 5, 10, 15, 20, 25, 30], [0, 5, 10, 15, 20, 25, 30])\n",
    "            ax3.set_yticks([0, 5, 10, 15, 20, 25, 30], [30, 25, 20, 15, 10, 5, 0])\n",
    "\n",
    "\n",
    "            # ax1.colorbar(im, fraction=0.02)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "            # seaborn.kdeplot(np.array(ml_delta), bw_method=0.1, ax=ax2)\n",
    "            delta_x, delta_y = diff_bar_graph_data(ml_delta)\n",
    "            ax2.bar(delta_x, delta_y, width=0.05)\n",
    "\n",
    "            # Updates\n",
    "            ax2.set_title(f\"Delta Methylation\")\n",
    "            ax2.set_xlabel(\"Methylation Diff\")\n",
    "            ax2.set_ylabel(\"# of CpG (in millions)\")\n",
    "\n",
    "            y0 = max(delta_y) * 1.02\n",
    "            ax2.plot([-0.43, 0.43], [y0, y0], color='#1f76b4',)\n",
    "            ax2.plot([0.15, 0.15], [0, y0], linestyle='dashed', color='#1f76b4',)\n",
    "            ax2.plot([-0.15, -0.15], [0, y0], linestyle='dashed', color='#1f76b4',)\n",
    "\n",
    "            t = [-0.45, -0.35, -0.25, -0.15, -0.05, 0.05, 0.15, 0.25, 0.35, 0.45]\n",
    "            tl = [-1, -0.35, -0.25, -0.15, -0.05, 0.05, 0.15, 0.25, 0.35, 1]\n",
    "            ax2.set_xticks(t, tl, rotation='vertical')\n",
    "\n",
    "            a = sum(delta_y[:3]) / sum(delta_y) * 100\n",
    "            b = sum(delta_y[3:6]) / sum(delta_y) * 100\n",
    "            c = sum(delta_y[6:]) / sum(delta_y) * 100\n",
    "\n",
    "            ax2.text(-0.35, y0*1.01, f\"{a:0.1f} %\", fontsize=10, color=\"#1f76b4\")\n",
    "            ax2.text(-0.05, y0*1.01, f\"{b:0.1f} %\", fontsize=10, color=\"#1f76b4\")\n",
    "            ax2.text( 0.25, y0*1.01, f\"{c:0.1f} %\", fontsize=10, color=\"#1f76b4\")\n",
    "\n",
    "            concordance = b\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "            fig4C.savefig(f\"./figures/S9_13.{pns}.pdf\", bbox_inches=\"tight\")\n",
    "\n",
    "            assert len(mg_cpg_cov) == len(mg_cpg_cov)\n",
    "            s1 = f\"\\t\\tY= {ml_k:0.4f}X + {ml_b:0.4f}, R^2 = {ml_r2:0.4f}, Pearson = {ml_r_pearson:0.4f}, N={len(mg_cpg_ml)/1e6:0.1f}M, concordance = {concordance:0.2f}%\"\n",
    "            s2 = f\"\\t\\tY= {cov_k:0.4f}X + {cov_b:0.4f}, R^2 = {cov_r2:0.4f}, Pearson = {cov_r_pearson:0.4f}, N={len(mg_cpg_cov)/1e6:0.1f}M\"\n",
    "\n",
    "            print(s1)\n",
    "            print(s2)\n",
    "\n",
    "            log_fh.write(pns + \"\\n\")\n",
    "            log_fh.write(f\"\\t{s}\\t{r}\\n\")\n",
    "            log_fh.write(s1 + \"\\n\")\n",
    "            log_fh.write(s2 + \"\\n\")\n",
    "            log_fh.flush()\n",
    "\n",
    "\n",
    "log_fh.close()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "f7bfd0f020932fa3",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [
    "\n",
    "\n",
    "replicate_comparison_str = f\"\"\"\n",
    "bs\n",
    "\t HG00621\n",
    "\t\tY= 0.8769X + 0.0938, R^2 = 0.7724, Pearson = 0.8788, N=8.5M, concordance = 72.17%\n",
    "\t\tY= 0.4622X + 4.1579, R^2 = 0.1619, Pearson = 0.4024, N=26.7M\n",
    "\t HG00741\n",
    "\t\tY= 0.8889X + 0.0691, R^2 = 0.7742, Pearson = 0.8799, N=16.1M, concordance = 68.47%\n",
    "\t\tY= 0.3640X + 4.1688, R^2 = 0.1672, Pearson = 0.4089, N=27.5M\n",
    "\t HG01952\n",
    "\t\tY= 0.8577X + 0.0868, R^2 = 0.7520, Pearson = 0.8672, N=12.1M, concordance = 66.48%\n",
    "\t\tY= 0.5292X + 4.8450, R^2 = 0.2088, Pearson = 0.4569, N=26.9M\n",
    "\t HG01978\n",
    "\t\tY= 0.8629X + 0.0866, R^2 = 0.7707, Pearson = 0.8779, N=11.7M, concordance = 67.46%\n",
    "\t\tY= 0.6521X + 6.1592, R^2 = 0.2152, Pearson = 0.4639, N=26.7M\n",
    "\t HG03516\n",
    "\t\tY= 0.8963X + 0.0635, R^2 = 0.7989, Pearson = 0.8938, N=20.6M, concordance = 71.07%\n",
    "\t\tY= 0.5026X + 4.9468, R^2 = 0.2424, Pearson = 0.4923, N=27.6M\n",
    "bc\n",
    "\t HG00621\n",
    "\t\tY= 0.8500X + 0.1134, R^2 = 0.7246, Pearson = 0.8512, N=15.0M, concordance = 65.97%\n",
    "\t\tY= 0.4307X + 5.2311, R^2 = 0.1543, Pearson = 0.3928, N=27.4M\n",
    "\t HG00741\n",
    "\t\tY= 0.8763X + 0.0733, R^2 = 0.7483, Pearson = 0.8650, N=20.5M, concordance = 64.28%\n",
    "\t\tY= 0.3437X + 5.1144, R^2 = 0.1521, Pearson = 0.3901, N=27.6M\n",
    "\t HG01952\n",
    "\t\tY= 0.8493X + 0.0904, R^2 = 0.7341, Pearson = 0.8568, N=19.5M, concordance = 63.47%\n",
    "\t\tY= 0.4473X + 6.4740, R^2 = 0.1743, Pearson = 0.4175, N=27.5M\n",
    "\t HG01978\n",
    "\t\tY= 0.8562X + 0.0931, R^2 = 0.7517, Pearson = 0.8670, N=19.8M, concordance = 64.67%\n",
    "\t\tY= 0.5213X + 7.9744, R^2 = 0.1775, Pearson = 0.4213, N=27.4M\n",
    "\t HG03516\n",
    "\t\tY= 0.8919X + 0.0639, R^2 = 0.7925, Pearson = 0.8902, N=24.5M, concordance = 69.12%\n",
    "\t\tY= 0.4767X + 6.9044, R^2 = 0.2135, Pearson = 0.4620, N=27.7M\n",
    "bm\n",
    "\t HG00621\n",
    "\t\tY= 0.8523X + 0.1110, R^2 = 0.7287, Pearson = 0.8536, N=15.9M, concordance = 66.01%\n",
    "\t\tY= 0.4136X + 5.5640, R^2 = 0.1423, Pearson = 0.3772, N=27.4M\n",
    "\t HG00741\n",
    "\t\tY= 0.8784X + 0.0716, R^2 = 0.7518, Pearson = 0.8671, N=21.2M, concordance = 64.47%\n",
    "\t\tY= 0.3316X + 5.3908, R^2 = 0.1406, Pearson = 0.3749, N=27.6M\n",
    "\t HG01952\n",
    "\t\tY= 0.8556X + 0.0867, R^2 = 0.7428, Pearson = 0.8619, N=21.4M, concordance = 64.26%\n",
    "\t\tY= 0.4410X + 6.6710, R^2 = 0.1757, Pearson = 0.4192, N=27.6M\n",
    "\t HG01978\n",
    "\t\tY= 0.8608X + 0.0905, R^2 = 0.7594, Pearson = 0.8714, N=21.8M, concordance = 65.44%\n",
    "\t\tY= 0.5066X + 8.3419, R^2 = 0.1765, Pearson = 0.4201, N=27.5M\n",
    "\t HG03516\n",
    "\t\tY= 0.8949X + 0.0634, R^2 = 0.7983, Pearson = 0.8935, N=25.2M, concordance = 69.67%\n",
    "\t\tY= 0.4550X + 7.6022, R^2 = 0.1937, Pearson = 0.4401, N=27.6M\n",
    "gb\n",
    "\t HG00621\n",
    "\t\tY= 0.8782X + 0.0917, R^2 = 0.7807, Pearson = 0.8836, N=3.3M, concordance = 70.81%\n",
    "\t\tY= 0.3919X + 3.8824, R^2 = 0.1013, Pearson = 0.3182, N=25.6M\n",
    "\t HG00741\n",
    "\t\tY= 0.8843X + 0.0710, R^2 = 0.7709, Pearson = 0.8780, N=11.1M, concordance = 67.33%\n",
    "\t\tY= 0.2973X + 4.0320, R^2 = 0.1068, Pearson = 0.3268, N=26.9M\n",
    "\t HG01952\n",
    "\t\tY= 0.8680X + 0.0797, R^2 = 0.7613, Pearson = 0.8725, N=9.7M, concordance = 66.62%\n",
    "\t\tY= 0.4187X + 4.5256, R^2 = 0.1407, Pearson = 0.3751, N=26.7M\n",
    "\t HG01978\n",
    "\t\tY= 0.8729X + 0.0794, R^2 = 0.7819, Pearson = 0.8842, N=8.7M, concordance = 67.62%\n",
    "\t\tY= 0.5130X + 5.6525, R^2 = 0.1412, Pearson = 0.3757, N=26.5M\n",
    "\t HG03516\n",
    "\t\tY= 0.8899X + 0.0674, R^2 = 0.7919, Pearson = 0.8899, N=16.4M, concordance = 69.88%\n",
    "\t\tY= 0.4163X + 4.7974, R^2 = 0.1657, Pearson = 0.4071, N=27.1M\n",
    "mg\n",
    "\tHG00621\n",
    "\t\tY= 0.8846X + 0.0883, R^2 = 0.7840, Pearson = 0.8854, N=14.8M, concordance = 73.60%\n",
    "\t\tY= 0.5048X + 4.6176, R^2 = 0.2104, Pearson = 0.4587, N=28.4M\n",
    "\n",
    "\tHG00741\n",
    "\t\tY= 0.8980X + 0.0636, R^2 = 0.7883, Pearson = 0.8879, N=20.2M, concordance = 70.03%\n",
    "\t\tY= 0.4216X + 4.1696, R^2 = 0.2306, Pearson = 0.4802, N=28.8M\n",
    "\n",
    "\tHG01952\n",
    "\t\tY= 0.8756X + 0.0763, R^2 = 0.7760, Pearson = 0.8809, N=20.2M, concordance = 69.16%\n",
    "\t\tY= 0.5421X + 5.4063, R^2 = 0.2563, Pearson = 0.5062, N=28.9M\n",
    "\n",
    "\tHG01978\n",
    "\t\tY= 0.8784X + 0.0790, R^2 = 0.7898, Pearson = 0.8887, N=20.7M, concordance = 69.99%\n",
    "\t\tY= 0.6327X + 6.8658, R^2 = 0.2611, Pearson = 0.5110, N=28.8M\n",
    "\n",
    "\tHG03516\n",
    "\t\tY= 0.9077X + 0.0573, R^2 = 0.8215, Pearson = 0.9064, N=24.7M, concordance = 73.57%\n",
    "\t\tY= 0.5645X + 5.7100, R^2 = 0.3003, Pearson = 0.5480, N=29.2M\n",
    "\n",
    "\"\"\".strip()\n",
    "\n",
    "\n",
    "method_comparison_str = \"\"\"\n",
    "bs\n",
    "\tHG00621\t1\n",
    "\t\tY= 0.9692X + 0.0229, R^2 = 0.9658, Pearson = 0.9827, N=10.9M, concordance = 96.30%\n",
    "\t\tY= 0.9242X + 2.2163, R^2 = 0.5941, Pearson = 0.7708, N=26.3M\n",
    "\tHG00621\t2\n",
    "\t\tY= 0.9713X + 0.0219, R^2 = 0.9677, Pearson = 0.9837, N=14.9M, concordance = 96.90%\n",
    "\t\tY= 0.9126X + 2.2870, R^2 = 0.6282, Pearson = 0.7926, N=26.7M\n",
    "\tHG00741\t1\n",
    "\t\tY= 0.9743X + 0.0165, R^2 = 0.9697, Pearson = 0.9847, N=20.5M, concordance = 96.68%\n",
    "\t\tY= 0.8998X + 2.5798, R^2 = 0.6266, Pearson = 0.7916, N=27.1M\n",
    "\tHG00741\t2\n",
    "\t\tY= 0.9769X + 0.0148, R^2 = 0.9734, Pearson = 0.9866, N=17.5M, concordance = 97.07%\n",
    "\t\tY= 0.8948X + 2.0800, R^2 = 0.6387, Pearson = 0.7992, N=27.0M\n",
    "\tHG01952\t1\n",
    "\t\tY= 0.9539X + 0.0309, R^2 = 0.9518, Pearson = 0.9756, N=13.6M, concordance = 93.33%\n",
    "\t\tY= 0.9322X + 3.5325, R^2 = 0.5535, Pearson = 0.7440, N=26.5M\n",
    "\tHG01952\t2\n",
    "\t\tY= 0.9649X + 0.0240, R^2 = 0.9620, Pearson = 0.9808, N=18.5M, concordance = 95.30%\n",
    "\t\tY= 0.8910X + 3.3579, R^2 = 0.5908, Pearson = 0.7686, N=27.0M\n",
    "\tHG01978\t1\n",
    "\t\tY= 0.9548X + 0.0293, R^2 = 0.9525, Pearson = 0.9760, N=11.9M, concordance = 92.77%\n",
    "\t\tY= 0.9406X + 3.5877, R^2 = 0.5318, Pearson = 0.7292, N=26.2M\n",
    "\tHG01978\t2\n",
    "\t\tY= 0.9666X + 0.0233, R^2 = 0.9633, Pearson = 0.9815, N=21.3M, concordance = 95.41%\n",
    "\t\tY= 0.8749X + 4.0541, R^2 = 0.5921, Pearson = 0.7695, N=27.1M\n",
    "\tHG03516\t1\n",
    "\t\tY= 0.9701X + 0.0205, R^2 = 0.9653, Pearson = 0.9825, N=22.2M, concordance = 95.86%\n",
    "\t\tY= 0.8926X + 3.6391, R^2 = 0.6038, Pearson = 0.7770, N=27.1M\n",
    "\tHG03516\t2\n",
    "\t\tY= 0.9673X + 0.0231, R^2 = 0.9624, Pearson = 0.9810, N=21.9M, concordance = 95.30%\n",
    "\t\tY= 0.8909X + 4.0830, R^2 = 0.5896, Pearson = 0.7678, N=27.1M\n",
    "bc\n",
    "\tHG00621\t1\n",
    "\t\tY= 0.9949X + 0.0053, R^2 = 0.9860, Pearson = 0.9930, N=16.5M, concordance = 98.75%\n",
    "\t\tY= 0.8935X + 0.6233, R^2 = 0.7770, Pearson = 0.8815, N=26.8M\n",
    "\tHG00621\t2\n",
    "\t\tY= 0.9930X + 0.0070, R^2 = 0.9852, Pearson = 0.9926, N=19.2M, concordance = 98.74%\n",
    "\t\tY= 0.9018X + 0.7093, R^2 = 0.7828, Pearson = 0.8847, N=27.0M\n",
    "\tHG00741\t1\n",
    "\t\tY= 0.9952X + 0.0047, R^2 = 0.9860, Pearson = 0.9930, N=23.1M, concordance = 98.59%\n",
    "\t\tY= 0.9066X + 0.7746, R^2 = 0.7696, Pearson = 0.8773, N=27.2M\n",
    "\tHG00741\t2\n",
    "\t\tY= 0.9955X + 0.0041, R^2 = 0.9879, Pearson = 0.9939, N=20.6M, concordance = 98.78%\n",
    "\t\tY= 0.9044X + 0.6771, R^2 = 0.7750, Pearson = 0.8803, N=27.1M\n",
    "\tHG01952\t1\n",
    "\t\tY= 0.9898X + 0.0100, R^2 = 0.9792, Pearson = 0.9896, N=19.9M, concordance = 97.63%\n",
    "\t\tY= 0.8611X + 1.4749, R^2 = 0.7034, Pearson = 0.8387, N=27.0M\n",
    "\tHG01952\t2\n",
    "\t\tY= 0.9926X + 0.0084, R^2 = 0.9825, Pearson = 0.9912, N=22.4M, concordance = 98.11%\n",
    "\t\tY= 0.8971X + 1.0633, R^2 = 0.7656, Pearson = 0.8750, N=27.2M\n",
    "\tHG01978\t1\n",
    "\t\tY= 0.9911X + 0.0091, R^2 = 0.9800, Pearson = 0.9900, N=19.2M, concordance = 97.61%\n",
    "\t\tY= 0.8659X + 1.3575, R^2 = 0.7083, Pearson = 0.8416, N=26.9M\n",
    "\tHG01978\t2\n",
    "\t\tY= 0.9915X + 0.0096, R^2 = 0.9813, Pearson = 0.9906, N=24.1M, concordance = 97.92%\n",
    "\t\tY= 0.8977X + 1.3548, R^2 = 0.7587, Pearson = 0.8710, N=27.2M\n",
    "\tHG03516\t1\n",
    "\t\tY= 0.9942X + 0.0065, R^2 = 0.9829, Pearson = 0.9914, N=24.5M, concordance = 98.15%\n",
    "\t\tY= 0.9047X + 1.0206, R^2 = 0.7671, Pearson = 0.8758, N=27.2M\n",
    "\tHG03516\t2\n",
    "\t\tY= 0.9928X + 0.0089, R^2 = 0.9811, Pearson = 0.9905, N=24.5M, concordance = 97.91%\n",
    "\t\tY= 0.9041X + 1.1808, R^2 = 0.7667, Pearson = 0.8756, N=27.2M\n",
    "bm\n",
    "\tHG00621\t1\n",
    "\t\tY= 0.9969X + 0.0035, R^2 = 0.9864, Pearson = 0.9932, N=16.9M, concordance = 98.87%\n",
    "\t\tY= 0.8970X + 0.4076, R^2 = 0.7908, Pearson = 0.8892, N=26.9M\n",
    "\tHG00621\t2\n",
    "\t\tY= 0.9954X + 0.0048, R^2 = 0.9861, Pearson = 0.9930, N=19.6M, concordance = 98.86%\n",
    "\t\tY= 0.9024X + 0.4551, R^2 = 0.7929, Pearson = 0.8905, N=27.0M\n",
    "\tHG00741\t1\n",
    "\t\tY= 0.9972X + 0.0033, R^2 = 0.9870, Pearson = 0.9935, N=23.3M, concordance = 98.74%\n",
    "\t\tY= 0.9087X + 0.5221, R^2 = 0.7721, Pearson = 0.8787, N=27.2M\n",
    "\tHG00741\t2\n",
    "\t\tY= 0.9973X + 0.0029, R^2 = 0.9888, Pearson = 0.9944, N=20.9M, concordance = 98.90%\n",
    "\t\tY= 0.9057X + 0.4738, R^2 = 0.7793, Pearson = 0.8828, N=27.1M\n",
    "\tHG01952\t1\n",
    "\t\tY= 0.9950X + 0.0043, R^2 = 0.9834, Pearson = 0.9917, N=21.1M, concordance = 98.40%\n",
    "\t\tY= 0.8939X + 0.4717, R^2 = 0.7829, Pearson = 0.8848, N=27.1M\n",
    "\tHG01952\t2\n",
    "\t\tY= 0.9955X + 0.0038, R^2 = 0.9853, Pearson = 0.9926, N=22.9M, concordance = 98.56%\n",
    "\t\tY= 0.9091X + 0.5112, R^2 = 0.7870, Pearson = 0.8871, N=27.2M\n",
    "\tHG01978\t1\n",
    "\t\tY= 0.9952X + 0.0038, R^2 = 0.9836, Pearson = 0.9918, N=20.4M, concordance = 98.34%\n",
    "\t\tY= 0.8951X + 0.4278, R^2 = 0.7834, Pearson = 0.8851, N=27.0M\n",
    "\tHG01978\t2\n",
    "\t\tY= 0.9949X + 0.0044, R^2 = 0.9846, Pearson = 0.9923, N=24.6M, concordance = 98.44%\n",
    "\t\tY= 0.9156X + 0.5843, R^2 = 0.7824, Pearson = 0.8845, N=27.2M\n",
    "\tHG03516\t1\n",
    "\t\tY= 0.9967X + 0.0040, R^2 = 0.9846, Pearson = 0.9923, N=24.7M, concordance = 98.40%\n",
    "\t\tY= 0.9096X + 0.5993, R^2 = 0.7672, Pearson = 0.8759, N=27.2M\n",
    "\tHG03516\t2\n",
    "\t\tY= 0.9960X + 0.0046, R^2 = 0.9837, Pearson = 0.9918, N=24.8M, concordance = 98.29%\n",
    "\t\tY= 0.9091X + 0.5953, R^2 = 0.7695, Pearson = 0.8772, N=27.2M\n",
    "gb\n",
    "\tHG00621\t1\n",
    "\t\tY= 0.9457X + 0.0432, R^2 = 0.9429, Pearson = 0.9710, N=5.4M, concordance = 92.04%\n",
    "\t\tY= 1.0821X + 3.0529, R^2 = 0.4831, Pearson = 0.6950, N=25.6M\n",
    "\tHG00621\t2\n",
    "\t\tY= 0.9533X + 0.0376, R^2 = 0.9492, Pearson = 0.9743, N=11.1M, concordance = 94.02%\n",
    "\t\tY= 1.0663X + 2.6510, R^2 = 0.5814, Pearson = 0.7625, N=26.5M\n",
    "\tHG00741\t1\n",
    "\t\tY= 0.9536X + 0.0323, R^2 = 0.9510, Pearson = 0.9752, N=17.4M, concordance = 93.22%\n",
    "\t\tY= 1.0349X + 3.1474, R^2 = 0.5855, Pearson = 0.7652, N=26.9M\n",
    "\tHG00741\t2\n",
    "\t\tY= 0.9598X + 0.0278, R^2 = 0.9568, Pearson = 0.9782, N=14.2M, concordance = 93.98%\n",
    "\t\tY= 1.0116X + 2.5055, R^2 = 0.6025, Pearson = 0.7762, N=26.7M\n",
    "\tHG01952\t1\n",
    "\t\tY= 0.9483X + 0.0367, R^2 = 0.9423, Pearson = 0.9707, N=12.5M, concordance = 91.27%\n",
    "\t\tY= 1.1027X + 3.0240, R^2 = 0.5764, Pearson = 0.7592, N=26.5M\n",
    "\tHG01952\t2\n",
    "\t\tY= 0.9524X + 0.0339, R^2 = 0.9484, Pearson = 0.9738, N=16.4M, concordance = 92.66%\n",
    "\t\tY= 1.0763X + 3.1812, R^2 = 0.5989, Pearson = 0.7739, N=26.8M\n",
    "\tHG01978\t1\n",
    "\t\tY= 0.9466X + 0.0367, R^2 = 0.9414, Pearson = 0.9702, N=9.7M, concordance = 90.02%\n",
    "\t\tY= 1.1309X + 3.3252, R^2 = 0.5326, Pearson = 0.7298, N=26.2M\n",
    "\tHG01978\t2\n",
    "\t\tY= 0.9547X + 0.0331, R^2 = 0.9490, Pearson = 0.9742, N=19.7M, concordance = 92.73%\n",
    "\t\tY= 1.0885X + 3.7784, R^2 = 0.5985, Pearson = 0.7736, N=26.9M\n",
    "\tHG03516\t1\n",
    "\t\tY= 0.9501X + 0.0361, R^2 = 0.9463, Pearson = 0.9728, N=19.6M, concordance = 92.22%\n",
    "\t\tY= 1.0780X + 4.1794, R^2 = 0.5630, Pearson = 0.7503, N=26.9M\n",
    "\tHG03516\t2\n",
    "\t\tY= 0.9512X + 0.0358, R^2 = 0.9460, Pearson = 0.9726, N=19.7M, concordance = 92.12%\n",
    "\t\tY= 1.0916X + 4.2599, R^2 = 0.5667, Pearson = 0.7528, N=26.9M\"\"\".strip()\n",
    "\n",
    "with open(\"figures/S9_13_4E_4F.log\") as log_fh:\n",
    "    method_comparison_str = log_fh.read().strip()\n",
    "\n",
    "\n",
    "replicate_comparison_data_pearson = {}\n",
    "replicate_comparison_data_concordance = {}\n",
    "\n",
    "sample = None\n",
    "pns = None\n",
    "for l in replicate_comparison_str.split(\"\\n\"):\n",
    "    l = l.strip()\n",
    "    if l == \"\":\n",
    "        continue\n",
    "\n",
    "    ll = l.split(\",\")\n",
    "    for i in range(len(ll)):\n",
    "        ll[i] = ll[i].strip()\n",
    "\n",
    "    if len(ll) == 1:\n",
    "        if ll[0] in [\"bs\", \"bc\", \"bm\", \"gb\", \"mg\"]:\n",
    "            pns = ll[0]\n",
    "        elif ll[0].startswith(\"HG0\"):\n",
    "            sample = ll[0]\n",
    "        else:\n",
    "            raise RuntimeError\n",
    "\n",
    "        continue\n",
    "\n",
    "    # print(ll)\n",
    "    if len(ll) < 5:\n",
    "        continue\n",
    "    if ll[4] == \"\":\n",
    "        continue\n",
    "\n",
    "\n",
    "    pearson_sc = float(ll[2].split(\"=\")[1])\n",
    "    concordance = float(ll[4].split(\"=\")[1][:-1]) / 100\n",
    "    # print(pns, sample, pearson_sc, concordance)\n",
    "\n",
    "    if pns not in replicate_comparison_data_pearson:\n",
    "        replicate_comparison_data_pearson[pns] = {}\n",
    "        replicate_comparison_data_concordance[pns] = {}\n",
    "\n",
    "    # print(pns, sample, pearson_sc, concordance)\n",
    "    replicate_comparison_data_pearson[pns][sample] = pearson_sc\n",
    "    replicate_comparison_data_concordance[pns][sample] = concordance\n",
    "\n",
    "\n",
    "\n",
    "method_comparison_data_pearson = {}\n",
    "method_comparison_data_concordance = {}\n",
    "\n",
    "sample = None\n",
    "pns = None\n",
    "for l in method_comparison_str.split(\"\\n\"):\n",
    "    l = l.strip()\n",
    "    if l == \"\":\n",
    "        continue\n",
    "\n",
    "    ll = l.split(\",\")\n",
    "    for i in range(len(ll)):\n",
    "        ll[i] = ll[i].strip()\n",
    "\n",
    "    if len(ll) == 1:\n",
    "        lll = ll[0].strip().split()\n",
    "\n",
    "        if len(lll) == 1:\n",
    "            assert lll[0] in [\"bs\", \"bc\", \"bm\", \"gb\", \"mg\"]\n",
    "            pns = lll[0]\n",
    "            continue\n",
    "\n",
    "        if lll[0].startswith(\"HG0\"):\n",
    "            sample = lll[0]\n",
    "            replicate = int(lll[1])\n",
    "            continue\n",
    "\n",
    "\n",
    "        sample = ll[0]\n",
    "        replicate = int(ll[1])\n",
    "        continue\n",
    "\n",
    "    if len(ll) < 5:\n",
    "        continue\n",
    "    if ll[4] == \"\":\n",
    "        continue\n",
    "\n",
    "    # print(pns, sample, replicate, ll)\n",
    "    pearson_sc = float(ll[2].split(\"=\")[1])\n",
    "    concordance = float(ll[4].split(\"=\")[1][:-1]) / 100\n",
    "\n",
    "    if pns not in method_comparison_data_pearson:\n",
    "        method_comparison_data_pearson[pns] = {}\n",
    "        method_comparison_data_concordance[pns] = {}\n",
    "\n",
    "    method_comparison_data_pearson[pns][(sample, replicate)] = pearson_sc\n",
    "    method_comparison_data_concordance[pns][(sample, replicate)] = concordance\n",
    "\n",
    "\n",
    "for pns in method_comparison_data_pearson:\n",
    "    # print(pns)\n",
    "    for s in method_comparison_data_pearson[pns]:\n",
    "        # print(s, method_comparison_data_pearson[pns][s], method_comparison_data_concordance[pns][s])\n",
    "        pass\n",
    "    # print()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "fig4XXX, axes4XXX = plt.subplots(5, 1, figsize=(8, 25))\n",
    "# h space\n",
    "fig4XXX.subplots_adjust(hspace=0.5)\n",
    "\n",
    "\n",
    "ax1 = axes4XXX[0]\n",
    "ax2 = axes4XXX[1]\n",
    "ax3 = axes4XXX[2]\n",
    "ax4 = axes4XXX[3]\n",
    "ax5 = axes4XXX[4]\n",
    "\n",
    "# 5 methods * 5 samples\n",
    "hmd_replicate_pearson = numpy.zeros((5, 5))\n",
    "hmd_replicate_concordance = numpy.zeros((5, 5))\n",
    "\n",
    "\n",
    "for i, pns in enumerate(pipeline_name_short):\n",
    "    for j, sample in enumerate(samples):\n",
    "        hmd_replicate_pearson[i, j] = replicate_comparison_data_pearson[pns][sample]\n",
    "        hmd_replicate_concordance[i, j] = replicate_comparison_data_concordance[pns][sample]\n",
    "\n",
    "\n",
    "\n",
    "ax1.imshow(hmd_replicate_pearson, cmap=\"Reds\", vmin=0.6, vmax=1)\n",
    "ax1.set_title(\"Replicate Pearson Correlation\")\n",
    "ax1.set_xticks(numpy.arange(len(samples)))\n",
    "ax1.set_xticklabels(samples, rotation=45)\n",
    "ax1.set_yticks(numpy.arange(len(pipeline_name_short)))\n",
    "ax1.set_yticklabels(pipeline_name_full)\n",
    "ax1.grid(False)\n",
    "\n",
    "ax2.imshow(hmd_replicate_concordance, cmap=\"Reds\", vmin=0.6, vmax=1)\n",
    "ax2.set_title(\"Replicate Concordance\")\n",
    "ax2.set_xticks(numpy.arange(len(samples)))\n",
    "ax2.set_xticklabels(samples, rotation=45)\n",
    "ax2.set_yticks(numpy.arange(len(pipeline_name_short)))\n",
    "ax2.set_yticklabels(pipeline_name_full)\n",
    "ax2.grid(False)\n",
    "\n",
    "for i in range(len(pipeline_name_short)):\n",
    "    for j in range(len(samples)):\n",
    "        ax1.text(j, i, f\"{hmd_replicate_pearson[i, j]:.2f}\", ha=\"center\", va=\"center\", color=\"white\")\n",
    "        ax2.text(j, i, f\"{hmd_replicate_concordance[i, j]:.2f}\", ha=\"center\", va=\"center\", color=\"black\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "hmd_method_pearson = numpy.zeros((4, 10))\n",
    "hmd_method_concordance = numpy.zeros((4, 10))\n",
    "\n",
    "\n",
    "x_labels = []\n",
    "y_labels = []\n",
    "for i, pns in enumerate(pipeline_name_short):\n",
    "    if pns == \"mg\":\n",
    "        continue\n",
    "    j = -1\n",
    "    x_labels = []\n",
    "    for s in samples:\n",
    "        for r in replicates:\n",
    "            j += 1\n",
    "            x_labels.append(f\"{s}_{r}\")\n",
    "            # print(pns, s, r)\n",
    "            hmd_method_pearson[i, j] = method_comparison_data_pearson[pns][(s, r)]\n",
    "            hmd_method_concordance[i, j] = method_comparison_data_concordance[pns][(s, r)]\n",
    "\n",
    "\n",
    "for i in range(4):\n",
    "    pns = pipeline_name_short[i]\n",
    "    yl = f\"{pipeline_name_short[i]} vs mG\"\n",
    "    y_labels.append(yl)\n",
    "\n",
    "ax3.imshow(hmd_method_pearson, cmap=\"Reds\", vmin=0.6, vmax=1)\n",
    "ax3.set_title(\"Method Pearson Correlation\")\n",
    "ax3.set_xticks(numpy.arange(len(x_labels)))\n",
    "ax3.set_xticklabels(x_labels, rotation=90)\n",
    "ax3.set_yticks(numpy.arange(4))\n",
    "ax3.set_yticklabels(y_labels)\n",
    "ax3.grid(False)\n",
    "\n",
    "ax4.imshow(hmd_method_concordance, cmap=\"Reds\", vmin=0.6, vmax=1)\n",
    "ax4.set_title(\"Method Concordance\")\n",
    "ax4.set_xticks(numpy.arange(len(x_labels)))\n",
    "ax4.set_xticklabels(x_labels, rotation=90)\n",
    "ax4.set_yticks(numpy.arange(4))\n",
    "ax4.set_yticklabels(y_labels)\n",
    "ax4.grid(False)\n",
    "\n",
    "\n",
    "\n",
    "for i in range(len(pipeline_name_short)):\n",
    "    if i == 4:\n",
    "        continue\n",
    "    for j in range(len(x_labels)):\n",
    "        ax3.text(j, i, f\"{hmd_method_pearson[i, j]:.2f}\", ha=\"center\", va=\"center\", color=\"white\")\n",
    "        ax4.text(j, i, f\"{hmd_method_concordance[i, j]:.2f}\", ha=\"center\", va=\"center\", color=\"white\")\n",
    "\n",
    "\n",
    "\n",
    "# make color scale for ax5\n",
    "xxx = numpy.zeros((1, 5))\n",
    "xxx[0, 0] = 0.6\n",
    "xxx[0, 1] = 0.7\n",
    "xxx[0, 2] = 0.8\n",
    "xxx[0, 3] = 0.9\n",
    "xxx[0, 4] = 1\n",
    "\n",
    "# add color bar\n",
    "f5 = ax5.imshow(xxx, cmap=\"Reds\", vmin=0.6, vmax=1)\n",
    "plt.colorbar(ax5.imshow(xxx, cmap=\"Reds\", vmin=0.6, vmax=1), ax=ax5, orientation=\"vertical\")\n",
    "\n",
    "\n",
    "\n",
    "fig4XXX.savefig(\"./figures/S13.pdf\", dpi=300, bbox_inches=\"tight\")\n",
    "\n",
    "\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "9ceaa12042e9c42f",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "ad89819e76d47d7b",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [
    "# PCA Plot\n",
    "\n",
    "cpg_pca_data = \"4D.PCA.1.tsv\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "bismark_cpg_data = {}\n",
    "methylgrapher_lifted_cytosine_data = {}\n",
    "\n",
    "\n",
    "for s in samples:\n",
    "    bismark_cpg_data[s] = {}\n",
    "    methylgrapher_lifted_cytosine_data[s] = {}\n",
    "\n",
    "\n",
    "    for r in replicates:\n",
    "        print(s, r)\n",
    "\n",
    "\n",
    "        bismark_cpg_data[s][r] = {}\n",
    "        methylgrapher_lifted_cytosine_data[s][r] = {}\n",
    "\n",
    "        mg2l_fp = f\"/scratch/wzhang/projects/ggWGBS/methylgrapher_no_trim/surjection/{s}_{r}.methylc.gz\"\n",
    "        bs_fp = f\"/scratch/wzhang/projects/ggWGBS/data/bismark_hg38/{s}_{r}.cpg.methylc\"\n",
    "\n",
    "        for l in open(bs_fp):\n",
    "            l = l.strip().split(\"\\t\")\n",
    "            chrom, start, end, context, ml, strand, cov = l\n",
    "\n",
    "            start = int(start)\n",
    "            met = int(round(float(ml) * int(cov)))\n",
    "            cov = int(cov)\n",
    "\n",
    "            if chrom not in bismark_cpg_data[s][r]:\n",
    "                bismark_cpg_data[s][r][chrom] = {}\n",
    "\n",
    "            bismark_cpg_data[s][r][chrom][start] = (met, cov)\n",
    "\n",
    "        for l in gzip.open(mg2l_fp, \"rt\"):\n",
    "            l = l.strip().split(\"\\t\")\n",
    "            chrom, start, end, context, ml, strand, cov = l\n",
    "\n",
    "            start = int(start)\n",
    "            met = int(round(float(ml) * int(cov)))\n",
    "            cov = int(cov)\n",
    "\n",
    "            if chrom not in methylgrapher_lifted_cytosine_data[s][r]:\n",
    "                methylgrapher_lifted_cytosine_data[s][r][chrom] = {}\n",
    "\n",
    "            methylgrapher_lifted_cytosine_data[s][r][chrom][start] = (met, cov)\n",
    "\n",
    "cpg_pca_data_fh = open(cpg_pca_data, \"w\")\n",
    "cpg_ref_fp = \"/scratch/wzhang/ref/linear/hg38/bismark/cpg_sites.bed\"\n",
    "for l in open(cpg_ref_fp):\n",
    "    l = l.strip().split(\"\\t\")\n",
    "    chrom, start, end = l\n",
    "    start = int(start) -1\n",
    "    \n",
    "    pass_cov = True\n",
    "\n",
    "    bs_cpgs = []\n",
    "    mg_cpgs = []\n",
    "\n",
    "    for s in samples:\n",
    "        for r in replicates:\n",
    "            if chrom not in bismark_cpg_data[s][r] or chrom not in methylgrapher_lifted_cytosine_data[s][r]:\n",
    "                pass_cov = False\n",
    "                break\n",
    "                \n",
    "            met, cov = bismark_cpg_data[s][r][chrom].get(start, (0, 0))\n",
    "            if cov <= min_cov:\n",
    "                pass_cov = False\n",
    "                break\n",
    "            bs_cpgs.append(met/cov)\n",
    "\n",
    "\n",
    "            m1, c1 = methylgrapher_lifted_cytosine_data[s][r][chrom].get(start, (0, 0))\n",
    "            m2, c2 = methylgrapher_lifted_cytosine_data[s][r][chrom].get(start+1, (0, 0))\n",
    "            met = m1 + m2\n",
    "            cov = c1 + c2\n",
    "            if cov <= min_cov:\n",
    "                pass_cov = False\n",
    "                break\n",
    "            mg_cpgs.append(met/cov)\n",
    "\n",
    "\n",
    "    if not pass_cov:\n",
    "        continue\n",
    "\n",
    "\n",
    "\n",
    "    line = [chrom, start, end] + bs_cpgs + mg_cpgs\n",
    "    line_str = \"\\t\".join(map(str, line)) + \"\\n\"\n",
    "    cpg_pca_data_fh.write(line_str)\n",
    "    \n",
    "cpg_pca_data_fh.close()\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "2fac9be9d98cd84b",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [
    "# https://hgdownload.cse.ucsc.edu/goldenpath/hg38/bigZips/hg38.chrom.sizes\n",
    "\n",
    "genome_size_str = \"\"\"chr1\t248956422\n",
    "chr2\t242193529\n",
    "chr3\t198295559\n",
    "chr4\t190214555\n",
    "chr5\t181538259\n",
    "chr6\t170805979\n",
    "chr7\t159345973\n",
    "chrX\t156040895\n",
    "chr8\t145138636\n",
    "chr9\t138394717\n",
    "chr11\t135086622\n",
    "chr10\t133797422\n",
    "chr12\t133275309\n",
    "chr13\t114364328\n",
    "chr14\t107043718\n",
    "chr15\t101991189\n",
    "chr16\t90338345\n",
    "chr17\t83257441\n",
    "chr18\t80373285\n",
    "chr20\t64444167\n",
    "chr19\t58617616\n",
    "chrY\t57227415\n",
    "chr22\t50818468\n",
    "chr21\t46709983\"\"\"\n",
    "\n",
    "genome_size = {}\n",
    "for l in genome_size_str.split(\"\\n\"):\n",
    "    l = l.strip().split(\"\\t\")\n",
    "    genome_size[l[0]] = int(l[1])\n",
    "\n",
    "\n",
    "genome_bins = {}\n",
    "bin_size = 3000\n",
    "for chrom in sorted(genome_size.keys(), key=lambda x: int(x[3:] if x[3:] != \"X\" and x[3:] != \"Y\" else \"23\" if x[3:] == \"M\" else \"24\")):\n",
    "    size = genome_size[chrom]\n",
    "    \n",
    "    start, end = 0, bin_size\n",
    "    \n",
    "    while end < size:\n",
    "        genome_bins[bin_id] = (chrom, start, end)\n",
    "        start += bin_size\n",
    "        end += bin_size\n",
    "        \n",
    "        bin_id = f\"{chrom}_{start}\"\n",
    "    print(bin_id)\n",
    "    \n",
    "\n",
    "def get_bin_id(chrom, start):\n",
    "    res = f\"{chrom}_{start//bin_size*bin_size}\"\n",
    "    if res not in genome_bins:\n",
    "        if len(chrom) > 5:\n",
    "            print(chrom, start, None)\n",
    "        return None\n",
    "    return res\n",
    "\n",
    "\n",
    "genome_bins_data = {}\n",
    "for l in open(cpg_pca_data):\n",
    "    l = l.strip().split(\"\\t\")\n",
    "    \n",
    "    chrom, start, end = l[:3]\n",
    "    values = list(map(float, l[3:]))\n",
    "    \n",
    "    start = int(start)\n",
    "    end = int(end)\n",
    "    \n",
    "    # print(chrom, start, end, len(values))\n",
    "    \n",
    "    bin_id = get_bin_id(chrom, start)\n",
    "    if bin_id == None:\n",
    "        continue\n",
    "    \n",
    "    if bin_id not in genome_bins_data:\n",
    "        genome_bins_data[bin_id] = []\n",
    "    \n",
    "    genome_bins_data[bin_id].append(values)\n",
    "    "
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "7fdc77f73112654",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [
    "# Use Per-CpG \n",
    "\n",
    "\n",
    "common_cpg_data = {}\n",
    "for l in open(cpg_pca_data):\n",
    "    l = l.strip().split(\"\\t\")\n",
    "    \n",
    "    chrom, start, end = l[:3]\n",
    "    values = list(map(float, l[3:]))\n",
    "    \n",
    "    ave = sum(values) / len(values)\n",
    "    all_the_same = True\n",
    "    for v in values:\n",
    "        if abs(v - ave) > 0.00001:\n",
    "            all_the_same = False\n",
    "    if all_the_same:\n",
    "        # print(chrom, start, \"all the same, \", values)\n",
    "        continue\n",
    "    \n",
    "    cpg_id = f\"{chrom}_{start}\"\n",
    "    common_cpg_data[cpg_id] = values\n",
    "\n",
    "cpg_ids = list(common_cpg_data.keys())\n",
    "    \n",
    "\n",
    "\n",
    "bin_pca_data_fh = open(\"4D.PCA.2.CpG.tsv\", \"w\")\n",
    "i = 0\n",
    "for m in [\"Bismark\", \"methylGrapher\"]:\n",
    "    for s in samples:\n",
    "        for r in replicates:\n",
    "            line = [s, r, m]\n",
    "            \n",
    "            for cpg_id in cpg_ids:\n",
    "                line.append(common_cpg_data[cpg_id][i])\n",
    "            \n",
    "            line_str = \"\\t\".join(list(map(str, line))) + \"\\n\"\n",
    "            bin_pca_data_fh.write(line_str)\n",
    "            \n",
    "            \n",
    "            i += 1\n",
    "bin_pca_data_fh.close()\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "23b3b7abff2f637a",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [
    "# Use Binned genome\n",
    "bin_pca_data = \"4D.PCA.2.tsv\"\n",
    "\n",
    "genome_bins_data_averaged = {}\n",
    "for bin_id in genome_bins_data:\n",
    "    \n",
    "    \n",
    "    sum_bin_ml = [0] * 20\n",
    "    cpg_count = len(genome_bins_data[bin_id])\n",
    "    \n",
    "    for each_cpg in genome_bins_data[bin_id]:\n",
    "        \n",
    "        for i in range(20):\n",
    "            sum_bin_ml[i] += each_cpg[i]\n",
    "    \n",
    "    for i in range(20):\n",
    "        sum_bin_ml[i] /= cpg_count\n",
    "    \n",
    "    if cpg_count < 5:\n",
    "        pass\n",
    "        # print(bin_id, f\"cpg count ({cpg_count}) < 10\")\n",
    "        #continue\n",
    "    \n",
    "    ave = sum(sum_bin_ml) / 20\n",
    "    all_the_same = True\n",
    "    for i in range(20):\n",
    "        if abs(sum_bin_ml[i] - ave) > 0.0001:\n",
    "            all_the_same = False\n",
    "    if all_the_same:\n",
    "        print(bin_id, f\"all the same, {sum_bin_ml}\")\n",
    "        continue\n",
    "    \n",
    "    \n",
    "    genome_bins_data_averaged[bin_id] = sum_bin_ml\n",
    "\n",
    "\n",
    "bin_ids = list(genome_bins_data_averaged.keys())\n",
    "\n",
    "\n",
    "bin_pca_data_fh = open(bin_pca_data, \"w\")\n",
    "i = 0\n",
    "for m in [\"Bismark\", \"methylGrapher\"]:\n",
    "    for s in samples:\n",
    "        for r in replicates:\n",
    "            line = [s, r, m]\n",
    "            \n",
    "            for bin_id in bin_ids:\n",
    "                line.append(genome_bins_data_averaged[bin_id][i])\n",
    "            \n",
    "            line_str = \"\\t\".join(list(map(str, line))) + \"\\n\"\n",
    "            bin_pca_data_fh.write(line_str)\n",
    "            \n",
    "            \n",
    "            i += 1\n",
    "bin_pca_data_fh.close()\n",
    "            \n",
    "\n",
    "\n",
    "\n",
    "    "
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "939299e3f37fc4cc",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "df = pandas.read_csv(\"4D.PCA.2.CpG.tsv\", sep=\"\\t\", header=None)\n",
    "# print(df.shape)\n",
    "# print(df.head())\n",
    "\n",
    "\n",
    "labels = {}\n",
    "\n",
    "for i in range(20):\n",
    "    s = df.iloc[i, 0]\n",
    "    r = df.iloc[i, 1]\n",
    "    m = df.iloc[i, 2]\n",
    "    labels[i] = f\"{s}_{r}_{m}\"\n",
    "\n",
    "# print()\n",
    "\n",
    "df2 = df.iloc[:, 3:]\n",
    "# print(df2.shape)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "pca = PCA(n_components=2)\n",
    "\n",
    "principalComponents = pca.fit_transform(df2)\n",
    "\n",
    "\n",
    "# print(principalComponents.shape)\n",
    "# print(principalComponents)\n",
    "\n",
    "\n",
    "for i in range(principalComponents.shape[0]):\n",
    "    label = labels[i]\n",
    "    pc1 = principalComponents[i, 0]\n",
    "    pc2 = principalComponents[i, 1]\n",
    "    \n",
    "    \n",
    "    # print(label, pc1, pc2)\n",
    "    \n",
    "\n",
    "\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "9bd20a9cfd4e1eb5",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [
    "fig, ax = plt.subplots(figsize=(8, 8))\n",
    "# ax.scatter(principalComponents[:, 0], principalComponents[:, 1])\n",
    "\n",
    "sample_colors = ['#1f77b4', '#ff7f0e', '#2ca02c', '#d62728', '#9467bd']\n",
    "# replicate_marker = [\".\" , \",\" , \"o\" , \"v\" , \"^\" , \"<\", \">\"]\n",
    "\n",
    "markers = [\",\" , \"^\" ]\n",
    "\n",
    "pca_data = {}\n",
    "for i in range(principalComponents.shape[0]):\n",
    "    label = labels[i]\n",
    "    s, r, m = label.split(\"_\")\n",
    "    if s not in pca_data:\n",
    "        pca_data[s] = {}\n",
    "    if r not in pca_data[s]:\n",
    "        pca_data[s][r] = {}\n",
    "    pca_data[s][r][m] = (principalComponents[i, 0], principalComponents[i, 1])\n",
    "\n",
    "\n",
    "for s in sorted(pca_data.keys()):\n",
    "    for r in sorted(pca_data[s].keys()):\n",
    "        for m in pca_data[s][r]:\n",
    "            \n",
    "            new_m = \"Bis\"\n",
    "            if m == \"methylGrapher\":\n",
    "                new_m = \"mG\"\n",
    "            new_label = f\"{s}_{r} ({new_m})\"\n",
    "            pc1 = pca_data[s][r][m][0]\n",
    "            pc2 = pca_data[s][r][m][1]\n",
    "            \n",
    "            color = sample_colors[samples.index(s)]\n",
    "            marker = markers[replicates.index(int(r))]\n",
    "            alpha = 1\n",
    "            \n",
    "            if m != \"Bismark\":\n",
    "                # marker = markers[replicates.index(int(r))+2]\n",
    "                alpha = 0.5\n",
    "            \n",
    "            ax.scatter([pc1], [pc2], color=color, label=new_label, marker=marker, alpha=alpha, s=60, facecolors='none')\n",
    "            \n",
    "            ax.set_xlabel(f'PC 1, Variance ({round(pca.explained_variance_ratio_[0]*100, 2)}%)')\n",
    "            ax.set_ylabel(f'PC 2, Variance ({round(pca.explained_variance_ratio_[1]*100, 2)}%)')\n",
    "            ax.set_title(f'Common CpG Site Methylation PCA (N = {df.shape[1]/1e6:0.2f}M)')\n",
    "\n",
    "ax.legend()\n",
    "plt.savefig(\"./figures/4D.PCA.CpG.pdf\", bbox_inches='tight')\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "184df231834a0e4b",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "bc1f62b77144a3df",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [
    "# Figure 5A\n",
    "\n",
    "#bismark_cpg_data = {}\n",
    "#methylgrapher_cytosine_data = {}\n",
    "\n",
    "pan_cpg_list = []\n",
    "pan_cpg_list_fp = f\"/scratch/wzhang/projects/ggWGBS/cpg_categorization/pan_v1_cpg.tsv\"\n",
    "hg38_loci_regex = re.compile(r\"hg38\\((.*):(\\d*)\\)\")\n",
    "\n",
    "\n",
    "seen = set()\n",
    "with open(pan_cpg_list_fp) as fh:\n",
    "    for l in fh:\n",
    "        l = l.strip().split(\"\\t\")\n",
    "        cpgid, segID1, seg_pos1, segID2, seg_pos2, cpg_type = l\n",
    "        seg_pos1 = int(seg_pos1)\n",
    "        seg_pos2 = int(seg_pos2)\n",
    "\n",
    "        info = tuple([cpg_type])\n",
    "        if cpg_type.startswith(\"hg38\"):\n",
    "            chrom, start = hg38_loci_regex.findall(cpg_type)[0]\n",
    "            start = int(start)\n",
    "            info = (\"hg38\", chrom, start)\n",
    "            # print(cpg_type, chrom, start)\n",
    "            # break\n",
    "\n",
    "        xxx = info[0]\n",
    "        if xxx not in seen:\n",
    "            seen.add(xxx)\n",
    "            print(xxx, l)\n",
    "\n",
    "        c = (segID1, seg_pos1, segID2, seg_pos2, info)\n",
    "        pan_cpg_list.append(c)\n",
    "        \n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "ea7dc9e318ae6582",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [
    "\n",
    "cpg_unique_data = {}\n",
    "\n",
    "for s in samples:\n",
    "    cpg_unique_data[s] = {}\n",
    "\n",
    "    for r in replicates:\n",
    "        print(s, r)\n",
    "        bs_cpg = bismark_cpg_data[s][r]\n",
    "        mg_cytosine = methylgrapher_cytosine_data[s][r]\n",
    "\n",
    "        total_bs = 0\n",
    "        for chrom in bs_cpg.keys():\n",
    "            total_bs += len(bs_cpg[chrom])\n",
    "\n",
    "        common, bss = 0, 0\n",
    "        mg_counts = [0, 0, 0]\n",
    "        for cpg in pan_cpg_list:\n",
    "            segID1, seg_pos1, segID2, seg_pos2, info = cpg\n",
    "\n",
    "            mge, bse = False, False\n",
    "            if segID1 in mg_cytosine:\n",
    "                if seg_pos1 in mg_cytosine[segID1]:\n",
    "                    mge = True\n",
    "            if segID2 in mg_cytosine:\n",
    "                if seg_pos2 in mg_cytosine[segID2]:\n",
    "                    mge = True\n",
    "\n",
    "            cpg_type = info[0]\n",
    "            if cpg_type == \"hg38\":\n",
    "                chrom = info[1]\n",
    "                start = info[2]\n",
    "                \n",
    "                if chrom in bs_cpg:\n",
    "                    if start in bs_cpg[chrom]:\n",
    "                        bse = True\n",
    "\n",
    "            if mge and bse:\n",
    "                common += 1\n",
    "            elif bse:\n",
    "                bss += 1\n",
    "            elif mge:\n",
    "                i = [\"hg38\", \"SNV\", \"SV\"].index(cpg_type)\n",
    "                #if i == 0 and len(chrom) > 5:\n",
    "                #     i = 3\n",
    "                mg_counts[i] += 1\n",
    "                \n",
    "        cpg_unique_data[s][r] = [bss, common] + mg_counts\n",
    "\n",
    "\n",
    "        \n",
    "        print(common)\n",
    "        print(mg_counts, sum(mg_counts))\n",
    "        print(bss)\n",
    "        print()\n",
    "\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "95b0e3d33266e41a",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [
    "\n",
    "fig5A, axes5A = plt.subplots(10, 1, figsize=(7, 28))\n",
    "\n",
    "axes_i = 0\n",
    "for s in samples:\n",
    "\n",
    "    for r in replicates:\n",
    "        d = cpg_unique_data[s][r][:]\n",
    "\n",
    "        bar_x = [\"Bis Unique\", \"Common\", \"mG within hg38\", \"mG SNV\", \"mG Others\"]\n",
    "\n",
    "        bar_x = list(reversed(bar_x))\n",
    "        d = list(reversed(d))\n",
    "        \n",
    "        axes5A[axes_i].barh(bar_x, d, color=\"#88a2d9\", height=0.6)\n",
    "\n",
    "        axes5A[axes_i].set_title(f\"{s}_{r}\")\n",
    "        # axes5A[axes_i].legend()\n",
    "\n",
    "        xticks = [0, 1e6, 2e6]\n",
    "        xticks_label = [0, 1, 2]\n",
    "        for i in range(5):\n",
    "            m = 0.5*1e7*(i+1)\n",
    "            label = str(round(5*(i+1)))\n",
    "            \n",
    "            xticks.append(m)\n",
    "            xticks_label.append(label)\n",
    "        \n",
    "        axes5A[axes_i].set_xticks(xticks, xticks_label)\n",
    "\n",
    "        axes_i += 1\n",
    "\n",
    "fig5A.savefig(f\"./figures/5A.pdf\", bbox_inches='tight')\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "be4e95d4dae2e166",
   "metadata": {},
   "source": [
    "\n",
    "for r in range(1, 7):\n",
    "    bs_hg38_url = f\"/scratch/wzhang/projects/ggWGBS/entex_dataset/bismark/ENC001/{r}/hg38/track.cpg.methylc.gz\"\n",
    "    mg_url = f\"/scratch/wzhang/projects/ggWGBS/methylgrapher_no_trim/ENC001/{r}/default.graph.CG.merged.tsv\"\n",
    "\n",
    "    c1, c2 = 0, 0\n",
    "    with gzip.open(bs_hg38_url, \"rt\") as fh:\n",
    "        for l in fh:\n",
    "            l = l.strip().split(\"\\t\")\n",
    "            cov = int(l[-1])\n",
    "            \n",
    "            if cov > 5:\n",
    "                c1 += 1\n",
    "            c2 +=1\n",
    "            \n",
    "    c3, c4 = 0, 0\n",
    "    with open(mg_url) as fh:\n",
    "        for l in fh:\n",
    "            l = l.strip().split(\"\\t\")\n",
    "            cov = int(l[-1])\n",
    "            \n",
    "            if cov > 5:\n",
    "                c3 += 1\n",
    "            c4 +=1\n",
    "    \n",
    "    print(r, c1, c2, c3, c4)\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "bb99d2b780b883ac",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [
    "figure6a_data_str = \"\"\"\n",
    "1 14616592 26865902 16429150 29074756\n",
    "2 15101570 26850694 17048709 29110264\n",
    "3 16563087 27364276 18779790 29703975\n",
    "4 13383780 26209787 16003132 28722950\n",
    "5 14706209 27334725 19019548 29989821\n",
    "6 11939444 25750073 14442447 28161547\n",
    "\"\"\".strip().split(\"\\n\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(5, 6))\n",
    "\n",
    "\n",
    "bismark_cpg_counts = []\n",
    "bismark_cpg_counts_cov5 = []\n",
    "methylgrapher_cpg_counts = []\n",
    "methylgrapher_cpg_counts_cov5 = []\n",
    "\n",
    "for l in figure6a_data_str:\n",
    "    l = l.strip().split(\" \")\n",
    "    for i in range(0, 5):\n",
    "        l[i] = int(l[i])\n",
    "\n",
    "    bismark_cpg_counts.append(l[2])\n",
    "    bismark_cpg_counts_cov5.append(l[1])\n",
    "\n",
    "    methylgrapher_cpg_counts.append(l[4])\n",
    "    methylgrapher_cpg_counts_cov5.append(l[3])\n",
    "\n",
    "    d = l[4] - l[2]\n",
    "    print(d, d/l[2])\n",
    "\n",
    "\n",
    "label = []\n",
    "bismark_x = []\n",
    "methylgrapher_x = []\n",
    "\n",
    "\n",
    "for i in range(6):\n",
    "\n",
    "    l = individual_experiment_tissue_label_short[\"ENC001\"][i+1]\n",
    "    lc = l[0].upper() + l[1:]\n",
    "    label.append(lc)\n",
    "    bismark_x.append(i-0.15)\n",
    "    methylgrapher_x.append(i+0.15)\n",
    "\n",
    "\n",
    "\n",
    "ax.bar(bismark_x, bismark_cpg_counts, width=0.3, color=pipeline_color_lookup[\"bs\"], label=\"Bismark (0<Cov<=5)\", alpha=0.5)\n",
    "ax.bar(methylgrapher_x, methylgrapher_cpg_counts, width=0.3, color=pipeline_color_lookup[\"mg\"], label=\"methylGrapher  (0<Cov<=5)\", alpha=0.5)\n",
    "\n",
    "ax.bar(bismark_x, bismark_cpg_counts_cov5, width=0.3, color=pipeline_color_lookup[\"bs\"], label=\"Bismark (Cov>5)\")\n",
    "ax.bar(methylgrapher_x, methylgrapher_cpg_counts_cov5, width=0.3, color=pipeline_color_lookup[\"mg\"], label=\"methylGrapher (Cov>5)\")\n",
    "\n",
    "ax.set_xticks(range(6), label)\n",
    "ax.set_title(\"# of CpG Site\")\n",
    "\n",
    "ax.legend(loc=\"lower right\")\n",
    "\n",
    "# rotate x label\n",
    "plt.xticks(rotation=90)\n",
    "\n",
    "\n",
    "ax.set_yticks([0, 1e7, 2e7, 3e7], [\"\", \"10M\", \"20M\", \"30M\"])\n",
    "\n",
    "plt.savefig(\"./figures/6A.pdf\", bbox_inches='tight')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "66e902e2aea13e34",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "9696d5fff748bddf",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [
    "# Divider - EN-TEx sample ENC001 analysis from now on"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "e45c447d5c2826c7",
   "metadata": {},
   "source": [
    "enc_001_long_read_vcf_fp = f\"/scratch/wzhang/projects/ggWGBS/genome_compare/vcf_analysis/ont.vcf\"\n",
    "\n",
    "ont_vcf_info = {}\n",
    "for i, l in enumerate(open(enc_001_long_read_vcf_fp)):\n",
    "    if l.startswith(\"#\"):\n",
    "        continue\n",
    "\n",
    "    l = l.strip().split(\"\\t\")\n",
    "    l[1] = int(l[1])\n",
    "\n",
    "    ont_vcf_info[i] = l\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "5ded22d75e848ca0",
   "metadata": {},
   "source": [
    "\n",
    "cpg_count = 0\n",
    "sv_lengths = []\n",
    "\n",
    "for ln, l in enumerate(open(enc_001_long_read_vcf_fp)):\n",
    "    if l.startswith(\"#\"):\n",
    "        continue\n",
    "\n",
    "    l = l.strip().split(\"\\t\")\n",
    "\n",
    "    chrom = l[0]\n",
    "    pos = int(l[1])\n",
    "    ref = l[3].upper()\n",
    "    alt = l[4].upper()\n",
    "\n",
    "    cpg_count += alt.upper().count(\"CG\")\n",
    "\n",
    "\n",
    "    if len(ref) != 1:\n",
    "        continue\n",
    "\n",
    "    if len(alt) <= 50:\n",
    "        continue\n",
    "\n",
    "    sv_lengths.append(len(alt))\n",
    "\n",
    "print(len(sv_lengths))\n",
    "print(sum(sv_lengths)/1e6)\n",
    "print(cpg_count)\n",
    "\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "2ecd00c401e4318e",
   "metadata": {},
   "source": [
    "\n",
    "figS16, axesS16 = plt.subplots(2, 1, figsize=(6, 13))\n",
    "plt.subplots_adjust(hspace=0.3)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "sv_perfect_match = []\n",
    "\n",
    "x = []\n",
    "y = []\n",
    "data2 = {}\n",
    "for ipc in [0.9, 0.91, 0.92, 0.93, 0.94, 0.95, 0.96, 0.97, 0.98, 0.985, 0.99, 0.992, 0.994, 0.996, 0.998, 0.999, 1]:\n",
    "    data2[ipc] = [[]]\n",
    "\n",
    "\n",
    "\n",
    "with open(\"/scratch/wzhang/projects/ggWGBS/genome_compare/vcf_analysis/flanking_mod/ins_50_flank_1000.graphalign.stat.tsv\") as fh:\n",
    "\n",
    "    for l in fh:\n",
    "        l = l.strip().split(\"\\t\")\n",
    "        svl = int(l[1])\n",
    "        ip = float(l[2])\n",
    "        # print(svl, ip)\n",
    "\n",
    "        x.append(svl)\n",
    "        y.append(ip)\n",
    "\n",
    "        if ip >= 1:\n",
    "            sv_perfect_match.append(svl)\n",
    "\n",
    "        for ipc in data2.keys():\n",
    "            if ip >= ipc:\n",
    "                data2[ipc][0].append(svl)\n",
    "\n",
    "\n",
    "\n",
    "x = list(sorted(data2.keys()))\n",
    "y1 = [len(data2[ipc][0]) for ipc in x]\n",
    "y2 = [sum(data2[ipc][0])/1e6 for ipc in x]\n",
    "\n",
    "\n",
    "axesS16[1].scatter(x, y1, s=1)\n",
    "axesS16[1].plot(x, y1)\n",
    "axesS16[1].set_title(\"#common insertions between pangenome and ENC001\")\n",
    "axesS16[1].set_xlabel(\"Identity percentage threshold\")\n",
    "axesS16[1].set_ylabel(\"# of insertions\")\n",
    "axesS16[1].set_ylim(0, 10000)\n",
    "axesS16[1].set_yticks(range(0, 10001, 2000))\n",
    "\n",
    "\n",
    "axesS16[0].scatter(x, y2, s=1)\n",
    "axesS16[0].plot(x, y2)\n",
    "axesS16[0].set_title(\"Total length common insertion between pangenome and ENC001\")\n",
    "axesS16[0].set_xlabel(\"Identity percentage threshold\")\n",
    "axesS16[0].set_ylabel(\"Total Insertion length (Mb)\")\n",
    "\n",
    "\n",
    "\n",
    "figS16.savefig(\"./figures/S16.pdf\", bbox_inches='tight')\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "94b5800205dc665f",
   "metadata": {},
   "source": [
    "\n",
    "flanking_length = 1000\n",
    "\n",
    "p_graph_fp = f\"/scratch/wzhang/projects/ggWGBS/genome_compare/vcf_analysis/flanking_mod/ins_50_flank_{flanking_length}.graphalign.stat.tsv\"\n",
    "p_h1_fp = f\"/scratch/wzhang/projects/ggWGBS/genome_compare/vcf_analysis/flanking_mod/ins_50_flank_{flanking_length}.hap1.minimap2.stat.tsv\"\n",
    "p_h2_fp = f\"/scratch/wzhang/projects/ggWGBS/genome_compare/vcf_analysis/flanking_mod/ins_50_flank_{flanking_length}.hap2.minimap2.stat.tsv\"\n",
    "\n",
    "\n",
    "graph_sv_lns = set()\n",
    "h1_sv_lns = set()\n",
    "h2_sv_lns = set()\n",
    "with open(p_graph_fp) as fh:\n",
    "    for l in fh:\n",
    "        l = l.strip().split()\n",
    "        ln = int(l[0].split(\":\")[0])\n",
    "        sv_length = int(l[1])\n",
    "        ip = float(l[2])\n",
    "        if ip < 0.9:\n",
    "            continue\n",
    "\n",
    "        graph_sv_lns.add(ln)\n",
    "        # print(ln, sv_length, ip)\n",
    "\n",
    "\n",
    "with open(p_h1_fp) as fh:\n",
    "    for l in fh:\n",
    "        l = l.strip().split()\n",
    "        ln = int(l[0].split(\":\")[0])\n",
    "        sv_length = int(l[1])\n",
    "        #ip = float(l[2])\n",
    "\n",
    "        if ln in graph_sv_lns:\n",
    "            h1_sv_lns.add(ln)\n",
    "\n",
    "with open(p_h2_fp) as fh:\n",
    "    for l in fh:\n",
    "        l = l.strip().split()\n",
    "        ln = int(l[0].split(\":\")[0])\n",
    "        sv_length = int(l[1])\n",
    "        #ip = float(l[2])\n",
    "\n",
    "        if ln in graph_sv_lns:\n",
    "            h2_sv_lns.add(ln)\n",
    "\n",
    "\n",
    "graph_sv_stat = [0, 0, 0]\n",
    "h1_sv_stat    = [0, 0, 0]\n",
    "h2_sv_stat    = [0, 0, 0]\n",
    "for ln in graph_sv_lns:\n",
    "\n",
    "    sv_seq = ont_vcf_info[ln][4].upper()\n",
    "    cg_count = sv_seq.count(\"CG\")\n",
    "    svl = len(sv_seq)\n",
    "\n",
    "    graph_sv_stat[0] += 1\n",
    "    graph_sv_stat[1] += svl\n",
    "    graph_sv_stat[2] += cg_count\n",
    "\n",
    "\n",
    "for ln in h1_sv_lns:\n",
    "\n",
    "    sv_seq = ont_vcf_info[ln][4].upper()\n",
    "    cg_count = sv_seq.count(\"CG\")\n",
    "    svl = len(sv_seq)\n",
    "\n",
    "    h1_sv_stat[0] += 1\n",
    "    h1_sv_stat[1] += svl\n",
    "    h1_sv_stat[2] += cg_count\n",
    "\n",
    "for ln in h2_sv_lns:\n",
    "\n",
    "    sv_seq = ont_vcf_info[ln][4].upper()\n",
    "    cg_count = sv_seq.count(\"CG\")\n",
    "    svl = len(sv_seq)\n",
    "\n",
    "    h2_sv_stat[0] += 1\n",
    "    h2_sv_stat[1] += svl\n",
    "    h2_sv_stat[2] += cg_count\n",
    "\n",
    "graph_sv_stat[1] /= 1e6\n",
    "h1_sv_stat[1] /= 1e6\n",
    "h2_sv_stat[1] /= 1e6\n",
    "\n",
    "print(graph_sv_stat)\n",
    "print(h1_sv_stat)\n",
    "print(h2_sv_stat)\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "b215ac49b2bfb977",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "flanking_length = 1000\n",
    "aligner = [\"graphalign\", \"minigraph\", \"vg\"][0]\n",
    "\n",
    "\n",
    "cpg_count = 0\n",
    "sv_count = 0\n",
    "sv_length_total = 0\n",
    "\n",
    "\n",
    "\n",
    "def methylc_reader(methylc_fp):\n",
    "    res = {}\n",
    "    with gzip.open(methylc_fp, \"rt\") as f:\n",
    "        for l in f:\n",
    "            l = l.strip().split(\"\\t\")\n",
    "            chrom, start, end, context, ml, strand, cov = l\n",
    "\n",
    "            start = int(start)\n",
    "            end = int(end)\n",
    "            ml = float(ml)\n",
    "            cov = int(cov)\n",
    "\n",
    "            if chrom not in res:\n",
    "                res[chrom] = {}\n",
    "            if start not in res[chrom]:\n",
    "                res[chrom][start] = (ml, cov)\n",
    "\n",
    "    return res\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "processed_fp = f\"/scratch/wzhang/projects/ggWGBS/genome_compare/vcf_analysis/flanking_mod/ins_50_flank_{flanking_length}.{aligner}.stat.tsv\"\n",
    "sv2graph = {}\n",
    "with open(processed_fp) as fh:\n",
    "    for l in fh:\n",
    "        l = l.strip().split(\"\\t\")\n",
    "        qn = l[0]\n",
    "        sv_length = int(l[1])\n",
    "        ip = float(l[2])\n",
    "        if ip < 0.9:\n",
    "            continue\n",
    "\n",
    "        coordinate_lifting_sv2graph = json.loads(l[4])\n",
    "\n",
    "        sv2graph[int(qn.split(\":\")[0])] = coordinate_lifting_sv2graph\n",
    "\n",
    "        alt_seq = ont_vcf_info[int(qn.split(\":\")[0])][4].upper()\n",
    "\n",
    "        cpg_count += alt_seq.count(\"CG\")\n",
    "        sv_count += 1\n",
    "        sv_length_total += len(alt_seq)\n",
    "\n",
    "\n",
    "\n",
    "sv2ind = {}\n",
    "for hi in [1, 2]:\n",
    "    processed_fp = f\"/scratch/wzhang/projects/ggWGBS/genome_compare/vcf_analysis/flanking_mod/ins_50_flank_{flanking_length}.hap{hi}.minimap2.stat.tsv\"\n",
    "\n",
    "    sv2ind[hi] = {}\n",
    "    with open(processed_fp) as fh:\n",
    "        for l in fh:\n",
    "            l = l.strip().split(\"\\t\")\n",
    "            qn = l[0]\n",
    "            sv2ind[hi][int(qn.split(\":\")[0])] = json.loads(l[3])\n",
    "\n",
    "\n",
    "sv_region = {}\n",
    "for hi in [1, 2]:\n",
    "    insertion_region_fp = f\"/scratch/wzhang/projects/ggWGBS/genome_compare/vcf_analysis/flanking_mod/ins_50_flank_{flanking_length}.hap{hi}.minimap2.insertion.bed\"\n",
    "    tl = 0\n",
    "    with open(insertion_region_fp) as fh:\n",
    "        for l in fh:\n",
    "            l = l.strip().split()\n",
    "            chrom, start, end = l\n",
    "            start = int(start)\n",
    "            end = int(end)\n",
    "            if chrom not in sv_region:\n",
    "                sv_region[chrom] = []\n",
    "            sv_region[chrom].append((start, end))\n",
    "            tl += end - start\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "individual = f\"ENC001\"\n",
    "\n",
    "for experiment_i in range(1, 7):\n",
    "\n",
    "    for hi in range(1, 3):\n",
    "\n",
    "        bs_hap = f\"/scratch/wzhang/projects/ggWGBS/entex_dataset/bismark/{individual}/{experiment_i}/hap{hi}/track.cpg.methylc.gz\"\n",
    "        bs_hap_data = methylc_reader(bs_hap)\n",
    "\n",
    "        bs_cpg_count = 0\n",
    "        for chrom in sv_region.keys():\n",
    "            for start, end in sv_region[chrom]:\n",
    "                for pos in range(start, end+1):\n",
    "                    if pos not in bs_hap_data.get(chrom, {}):\n",
    "                        continue\n",
    "                    bs_cpg_count += 1\n",
    "\n",
    "        mgl_hap = f\"/scratch/wzhang/projects/ggWGBS/methylgrapher_no_trim/{individual}/{experiment_i}/hap{hi}.cpg.methylc.gz\"\n",
    "        mgl_hap_data = methylc_reader(mgl_hap)\n",
    "\n",
    "        mgl_cpg_count = 0\n",
    "        for chrom in sv_region.keys():\n",
    "            for start, end in sv_region[chrom]:\n",
    "                for pos in range(start, end+1):\n",
    "                    if pos not in mgl_hap_data.get(chrom, {}):\n",
    "                        continue\n",
    "                    mgl_cpg_count += 1\n",
    "\n",
    "        print(individual, experiment_i, hi, bs_cpg_count, mgl_cpg_count)\n",
    "\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "39de1b4f8c748303",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "fig, axes = plt.subplots(6, 2, figsize=(8, 26))\n",
    "plt.subplots_adjust(hspace=0.3)\n",
    "\n",
    "def methylc_reader(methylc_fp):\n",
    "    res = {}\n",
    "    with gzip.open(methylc_fp, \"rt\") as f:\n",
    "        for l in f:\n",
    "            l = l.strip().split(\"\\t\")\n",
    "            chrom, start, end, context, ml, strand, cov = l\n",
    "\n",
    "            start = int(start)\n",
    "            end = int(end)\n",
    "            ml = float(ml)\n",
    "            cov = int(cov)\n",
    "\n",
    "            if chrom not in res:\n",
    "                res[chrom] = {}\n",
    "            if start not in res[chrom]:\n",
    "                res[chrom][start] = (ml, cov)\n",
    "\n",
    "    return res\n",
    "\n",
    "individual = f\"ENC001\"\n",
    "for experiment_i in range(1, 7):\n",
    "\n",
    "    tissue_name = individual_experiment_tissue_label_short[individual].get(experiment_i, \"Unknown\")\n",
    "    # Capitalize tissue name\n",
    "    tissue_name = tissue_name[0].upper() + tissue_name[1:]\n",
    "\n",
    "    for hi in range(1, 3):\n",
    "        ax = axes[experiment_i-1][hi-1]\n",
    "        bs_hap = f\"/scratch/wzhang/projects/ggWGBS/entex_dataset/bismark/{individual}/{experiment_i}/hap{hi}/track.cpg.methylc.gz\"\n",
    "        lifted = f\"/scratch/wzhang/projects/ggWGBS/methylgrapher_no_trim/ENC001/{experiment_i}/hap{hi}.cpg.methylc.gz\"\n",
    "\n",
    "        bs_hap_methylc = methylc_reader(bs_hap)\n",
    "        lifted_methylc = methylc_reader(lifted)\n",
    "\n",
    "        cpg_count = 0\n",
    "        x = []\n",
    "        y = []\n",
    "        for chrom in bs_hap_methylc:\n",
    "            if chrom not in lifted_methylc:\n",
    "                continue\n",
    "\n",
    "            for start in bs_hap_methylc[chrom]:\n",
    "                bs_ml, bs_cov = bs_hap_methylc[chrom][start]\n",
    "                lifted_met, lifted_cov = 0, 0\n",
    "\n",
    "\n",
    "                lifted_d0 = lifted_methylc[chrom].get(start, (0, 0))\n",
    "                lifted_ml = lifted_d0[0]\n",
    "                lifted_cov = lifted_d0[1]\n",
    "\n",
    "                if lifted_cov == 0:\n",
    "                    continue\n",
    "\n",
    "                cpg_count += 1\n",
    "                if bs_cov <= 5 or lifted_cov <= 5:\n",
    "                    continue\n",
    "\n",
    "                x.append(bs_ml)\n",
    "                y.append(lifted_ml)\n",
    "\n",
    "\n",
    "        k, b, r, p, std_err = scipy.stats.linregress(x, y)\n",
    "\n",
    "\n",
    "        concordance_interval = 0.15\n",
    "        concordance = 0\n",
    "        for i in range(len(x)):\n",
    "            x0 = x[i]\n",
    "            y0 = y[i]\n",
    "\n",
    "            cflag = abs(x0 - y0) <= concordance_interval\n",
    "            if cflag:\n",
    "                concordance += 1\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        ax.scatter(x, y, s=0.8)\n",
    "        ax.plot([0, 1], [b, k+b], color=\"green\")\n",
    "        ax.text(0.08, 0.75, f\"y = {k:.2f}x + {b:.2f}\\nr = {r:.2f}\\nN = {len(x)}\", transform=ax.transAxes)\n",
    "\n",
    "        ax.set_title(f\"{tissue_name} hap{hi}\")\n",
    "        ax.set_xlabel(\"Methylation (Bismark)\")\n",
    "        ax.set_ylabel(\"Methylation (methylGrapher)\")\n",
    "\n",
    "        print(individual, experiment_i, hi, cpg_count, len(x), k, b, r, concordance/len(x))\n",
    "\n",
    "\n",
    "        fig.savefig(f\"./figures/S17.pdf\", dpi=300, bbox_inches='tight')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "2a659f2d76188fa7",
   "metadata": {},
   "source": [],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "1ce70fb4b7a195db",
   "metadata": {},
   "source": [],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "bac546b94e79d4bd",
   "metadata": {},
   "source": [],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "eb84d92ccbebf334",
   "metadata": {},
   "source": [],
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
